{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is the [original notebook](https://github.com/malariagen/PDNA_SP_markers/blob/9b73a05d2c573e66ea70b282adb7055a66e2022e/work/3_haplotype_plots/20221221_haplotype_plots.ipynb) Richard wanted us to update and use:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 21 19:06:20 GMT 2022\n",
      "node-10-1-4\n",
      "/nfs/team112_data03/personal/rp7/src/github/malariagen/PDNA_SP_markers/work/3_haplotype_plots\n"
     ]
    }
   ],
   "source": [
    "!date\n",
    "!hostname\n",
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       ".container {\n",
       "    width: 100%;\n",
       "}\n",
       "# div#notebook {\n",
       "#     padding-top: 1em;\n",
       "# }\n",
       "# #header-container {\n",
       "#     display: none;\n",
       "# }\n",
       "# #header-bar {\n",
       "#     display: none;\n",
       "# }\n",
       "# #maintoolbar {\n",
       "#     display: none;\n",
       "# }\n",
       "# #menubar-container {\n",
       "#     position: fixed;\n",
       "#     margin-top: 0;\n",
       "# }\n",
       "site {\n",
       "    height: auto !important;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python 3.8.6 (default, Oct 13 2020, 20:37:26) \n",
      "[GCC 8.3.0]\n",
      "numpy 1.21.4\n",
      "scipy 1.4.1\n",
      "pandas 1.2.3\n",
      "pysam 0.15.4\n",
      "petlx 1.0.3\n",
      "vcf 0.6.8\n",
      "h5py 2.10.0\n",
      "zarr 2.4.0\n",
      "scikit-allel 1.2.1\n"
     ]
    }
   ],
   "source": [
    "%run ../setup_abox.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inputs\n",
    "vcf_file_format = \"/lustre/scratch124/gsu/legacy/pipelines/builds/pf_70_build/pf_70_internal_release/vcf/%s.pf7.vcf.gz\" # Location of VCFs, allowing chrom to vary\n",
    "samples_fn = '/lustre/scratch126/gsu/team112/pf7/ftp_20221021/Pf7_samples.txt'\n",
    "ref_genome_fn = '/lustre/scratch124/gsu/legacy/pfalciparum/resources/Pfalciparum.genome.fasta'\n",
    "gff_fn = '/lustre/scratch124/gsu/legacy/pfalciparum/resources/snpEff/data/Pfalciparum_GeneDB_Feb2020/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "population_colours = collections.OrderedDict()\n",
    "# population_colours['SA']     = \"#4daf4a\"\n",
    "population_colours['AF-W']= \"#e31a1c\"\n",
    "population_colours['AF-C'] = \"#fd8d3c\" \n",
    "population_colours['AF-NE'] = \"#bb8129\" \n",
    "population_colours['AF-E'] = \"#fecc5c\"\n",
    "# population_colours['AS-S-E'] = \"#dfc0eb\" \n",
    "# population_colours['AS-S-FE'] = \"#984ea3\" \n",
    "# population_colours['AS-SE-W'] = \"#9ecae1\"\n",
    "# population_colours['AS-SE-E'] = \"#3182bd\"\n",
    "# population_colours['OC-NG'] = \"#f781bf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# locations with at least 25 samples (see 20210218_drm_frequency_over_time_pops_adminlevel1.ipynb) (borderline exceptions are marked on this notebook)\n",
    "# I think we should be able to generate this with code, but for now this is fine\n",
    "locations = [\n",
    "\n",
    "#     ('SA', 'Colombia', 'Cauca'),\n",
    "\n",
    "    ('AF-W', 'Gabon', 'Wouleu-Ntem'),\n",
    "    ('AF-W', 'Cameroon', 'Sud-Ouest'),\n",
    "    ('AF-W', 'Nigeria', 'Lagos'),\n",
    "    ('AF-W', 'Benin', 'Littoral'),\n",
    "    ('AF-W', 'Benin', 'Atlantique'),\n",
    "    ('AF-W', 'Ghana', 'Volta'),\n",
    "    ('AF-W', 'Ghana', 'Greater Accra'),\n",
    "    ('AF-W', 'Ghana', 'Upper East'),\n",
    "    ('AF-W', 'Ghana', 'Central'),\n",
    "    ('AF-W', 'Ghana', 'Ashanti'),\n",
    "    ('AF-W', 'Ghana', 'Brong Ahafo'),\n",
    "    ('AF-W', 'Burkina Faso', 'Haut-Bassins'),\n",
    "    ('AF-W', 'Mali', 'Segou'),\n",
    "    ('AF-W', 'Mali', 'Sikasso'),\n",
    "    ('AF-W', 'Mali', 'Koulikoro'),\n",
    "    ('AF-W', 'Mali', 'Bamako'),\n",
    "    ('AF-W', 'Mali', 'Kayes'),\n",
    "    \n",
    "    \n",
    "    ('AF-W', \"CÃ´te d'Ivoire\", 'Abidjan'),\n",
    "    ('AF-W', 'Mauritania', 'Hodh el Gharbi'),\n",
    "    ('AF-W', 'Mauritania', 'Hodh ech Chargui'), \n",
    "    ('AF-W', 'Guinea', 'Nzerekore'),\n",
    "    ('AF-W', 'Guinea', 'Faranah'),\n",
    "    ('AF-W', 'Senegal', 'Sedhiou'),\n",
    "    ('AF-W', 'Senegal', 'Dakar'),\n",
    "    ('AF-W', 'Gambia', 'Upper River'),\n",
    "    ('AF-W', 'Gambia', 'North Bank'),\n",
    "    ('AF-W', 'Gambia', 'Western'),\n",
    "    \n",
    "    ('AF-C', 'DRC', 'Kinshasa'),\n",
    "\n",
    "    ('AF-NE', 'Kenya', 'Kisumu'),\n",
    "    ('AF-NE', 'Sudan', 'Khartoum'),\n",
    "\n",
    "    ('AF-E', 'Kenya', 'Kilifi'),\n",
    "    ('AF-E', 'Mozambique', 'Gaza'), \n",
    "    ('AF-E', 'Tanzania', 'Lindi'),\n",
    "    ('AF-E', 'Tanzania', 'Tanga'),\n",
    "    ('AF-E', 'Tanzania', 'Morogoro'),\n",
    "    ('AF-E', 'Tanzania', 'Kagera'),\n",
    "    ('AF-E', 'Tanzania', 'Kigoma'),\n",
    "    ('AF-E', 'Malawi', 'Zomba'),   \n",
    "    ('AF-E', 'Malawi', 'Chikwawa'),\n",
    "    \n",
    "#     ('AS-S-E', 'India', 'West Bengal'),\n",
    "#     ('AS-S-E', 'India', 'Odisha'),\n",
    "\n",
    "#     ('AS-S-FE', 'Bangladesh', 'Chittagong'),\n",
    "#     ('AS-S-FE', 'India', 'Tripura'),\n",
    "\n",
    "#     ('AS-SE-W', 'Thailand', 'Tak'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Tanintharyi'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Shan'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Kayin'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Kachin'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Bago'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Mandalay'),\n",
    "#     ('AS-SE-W', 'Myanmar', 'Sagaing'),\n",
    "\n",
    "#     ('AS-SE-E', 'Vietnam', 'Khanh Hoa'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Ninh Thuan'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Gia Lai'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Dak Lak'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Quang Nam'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Dak Nong'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Quang Tri'),\n",
    "#     ('AS-SE-E', 'Vietnam', 'Binh Phuoc'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Ratanakiri'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Stueng Traeng'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Preah Vihear'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Pursat'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Battambang'),\n",
    "#     ('AS-SE-E', 'Cambodia', 'Pailin'),\n",
    "    \n",
    "#     ('AS-SE-E', 'Laos', 'Sekong'),\n",
    "#     ('AS-SE-E', 'Laos', 'Attapeu'),\n",
    "#     ('AS-SE-E', 'Laos', 'Salavan'),\n",
    "#     ('AS-SE-E', 'Laos', 'Champasak'),\n",
    "#     ('AS-SE-E', 'Laos', 'Savannakhet'),\n",
    "#     ('AS-SE-E', 'Thailand', 'Sisakhet'),\n",
    "\n",
    "#     ('OC-NG', 'Papua New Guinea', 'Milne Bay'),\n",
    "#     ('OC-NG', 'Papua New Guinea', 'Madang'),\n",
    "#     ('OC-NG', 'Papua New Guinea', 'East Sepik'),\n",
    "#     ('OC-NG', 'Indonesia', 'Papua')\n",
    "]       \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene_name': 'CRT',\n",
       " 'chrom': 'Pf3D7_07_v3',\n",
       " 'strand': '+',\n",
       " 'starts': array([403222, 403490, 403938, 404283, 404569, 404764, 404936, 405146,\n",
       "        405334, 405539, 405825, 406017, 406241]),\n",
       " 'ends': array([403312, 403758, 404110, 404415, 404640, 404839, 405018, 405196,\n",
       "        405390, 405631, 405869, 406071, 406317])}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def determine_cds(gene_id='PF3D7_0709000', gff_fn=gff_fn):\n",
    "    df_gff = allel.gff3_to_dataframe(gff_fn, attributes=['ID', 'Name'])\n",
    "    gene_name = df_gff.loc[df_gff['ID'] == gene_id, 'Name'].values[0]\n",
    "    if (gene_name == '') or (gene_name == '.'):\n",
    "        gene_name=gene_id\n",
    "    chrom = df_gff.loc[df_gff['ID'] == gene_id, 'seqid'].values[0]\n",
    "    strand = df_gff.loc[df_gff['ID'] == gene_id, 'strand'].values[0]\n",
    "    df_cds = df_gff.loc[\n",
    "        ( df_gff['ID'].str.startswith(gene_id) )\n",
    "        & ( df_gff['type'] == 'CDS' )\n",
    "    ]\n",
    "    return(\n",
    "        {\n",
    "            'gene_name':gene_name,\n",
    "            'chrom':chrom,\n",
    "            'strand':strand,\n",
    "            'starts':df_cds['start'].values,\n",
    "            'ends':df_cds['end'].values,\n",
    "        }\n",
    "    )\n",
    "\n",
    "determine_cds()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# call_haplotype function\n",
    "####################################\n",
    "\n",
    "# This function creates nucleotide and amino acid haplotypes of specific regions, for example:\n",
    "# 1) A single amino acid such as CRT 76\n",
    "# 2) A stretch of contiguous amino acids such as CRT 72-76\n",
    "# 3) The full coding region of a gene such as CRT\n",
    "\n",
    "# Note that in the case of 3), if the gene has multiple exons, the region of interest will be non-contiguous\n",
    "\n",
    "# The function can be used to call a) a single sample, b) specific samples of interest or c) all samples in the VCF (the default).\n",
    "# Note that calling all samples isn't much slower than calling a single sample. Calling all haplotypes for all Pf7 samples took ~40 minutes\n",
    "\n",
    "# The function returns a tuple with three elements:\n",
    "# 1) A pandas DataFrame with sample ID as index, and three columns aa_haplotype, nucleotide_haplotype and ns_changes\n",
    "# 2) The (3D7) reference amino acid haplotype\n",
    "# 3) The (3D7) reference nucleotide haplotype\n",
    "\n",
    "# Note that the reference haplotypes are probably not required. They were largely intended for debugging/sanity checking purposes\n",
    "# As such the key output is the first element (the DataFrame)\n",
    "\n",
    "####################################\n",
    "# Explanation of column aa_haplotype\n",
    "####################################\n",
    "# If the nucleotide haplotype is homozygous (i.e. there are no heterozygous variants), there will be a single amino acid haplotype\n",
    "# If the nucleotide haplotype is heterozygous (i.e. there is at least one heterozygous variants), there will be two amino acid haplotypes, separated by a comma\n",
    "# Note that in some cases, there could be a heterozygous nucleotide haplotype, both alleles of which give the same amino acid haplotype.\n",
    "# In cases such as this, the amino acid haplotype will be repeated, separated by a comma \n",
    "# If there are multiple mutations that are heterozygous, and cannot be phased, an asterisk (*) will be appended to the aa_haplotype call.\n",
    "# If any variant is missing, aa_haplotype will be '-'.\n",
    "# If the resulting nucleotide haplotype has a length that is not a multiple of 3 (due to indels) the aa_haplotype will be \"!\" to represent a frameshift\n",
    "# In my experience to date, such frameshifts have only be seen as heterozygote calls, i.e. \"<haplotype>,!\" or \"!,<haplotype>\", though would\n",
    "# probably be much more common if setting `include_indels=False`\n",
    "\n",
    "####################################\n",
    "# Explanation of column nucleotide_haplotype\n",
    "####################################\n",
    "# If the nucleotide haplotype is homozygous (i.e. there are no heterozygous variants), there will be a single haplotype\n",
    "# If the nucleotide haplotype is heterozygous (i.e. there is at least one heterozygous variants), there will be two haplotypes, separated by a comma\n",
    "# Nucleotides which match the reference sequence are shown in lower case, whereas nucleotides that differ (SNPs) are shown in upper case\n",
    "# If any variant is missing, nucleotide_haplotype will be '-'.\n",
    "\n",
    "####################################\n",
    "# Explanation of column ns_changes\n",
    "####################################\n",
    "# All non-synonymous mutations will be shown in the form \"<REF AA><AA number><ALT AA>\", e.g. \"K76T\" or \"C580Y\"\n",
    "# Homozygous mutations are shown in upper case and heterozygous in lower case.\n",
    "# If the code finds two distinct amino acid haplotypes, these will be separated by a comma.\n",
    "# If multiple NS mutations are found within an amino acid haplotype, these will be separated by a forward slash (/).\n",
    "# If there are multiple mutations that are heterozygous, and cannot be phased, an asterisk (*) will be appended to the ns_changes call.\n",
    "# Note that this still holds\n",
    "# If there are no NS mutations, ns_changes will be NaN, with the exception of 2 or more unphased synonymous variants, in which case ns_changes will be '*'\n",
    "# This use of '*' to represent multiple unphased synonymous mutations is perhaps undesirable and we might want to review this in the future\n",
    "# Missing variants will be ignored, but any NS mutation elsewhere will still be included. Note the consequence of this is that it is\n",
    "# possible that aa_haplotype will be '-' but ns_changes will have one or more NS mutations reported.\n",
    "# If the resulting nucleotide haplotype has a length that is not a multiple of 3 (due to indels) ns_changes will be \"!\" to represent a frameshift\n",
    "# If any mutations result in a premature stop codon, ns_changes will also be \"!\"\n",
    "\n",
    "####################################\n",
    "# Outline of the code\n",
    "####################################\n",
    "# This function is complex and difficult to follow (I had trouble coming back to it a few months after initially\n",
    "# writing it). In essence this determines the reference nucleotide sequence of the region (ref_sequence), and then\n",
    "# for each sample applies the mutations for the first and second aa_haplotype calls (remembering there are two calls as\n",
    "# we are calling as diploid with GATK). This takes account of complications such as phased het calls and indels changing\n",
    "# the length of the sequence (though indels are not included by default). The reference and two alternative nucleotide\n",
    "# sequences are then converted to amino acid sequences, and mutations detected.\n",
    "\n",
    "####################################\n",
    "# Known issues\n",
    "####################################\n",
    "# Note that if using the default option of `include_indels=False` we shouldn't really see any frameshifts (represented by \"!\" in aa_haplotype),\n",
    "# however, these are occasionally seen in the following two circumstances:\n",
    "# 1) By default we allow the 1bp insertion and deletion at Pf3D7_07_v3:403618 and Pf3D7_07_v3:403622\n",
    "#   (due to default option `positions_to_allow=[403618, 403622]`). These are usually seen together, so result in in-frame changes\n",
    "#   However, in a few cases only one of the indels is called resulting in a frameshift\n",
    "# 2) In some cases a SNP is represented by GATK as REF and ALT alleles longer than 1 but only differing in the first position.\n",
    "#    An example of this in Pf7 is `Pf3D7_08_v3\t549683\t.\tTGG\tCGG,*`. There is a bug in the code that I haven't had chance to look into\n",
    "#    which results in samples with this SNP being represented as a frameshift\n",
    "#\n",
    "# It is possible (and in fact quite common) to have a missing aa_haplotype (represented by \"-\"), but to have one or more mutations in ns_changes.\n",
    "# This was by design and intended specifically for the gene Kelch13 where if we see any non-synonymous mutation between AA 349 and 726\n",
    "# we can assign a resistant phenotype, even if there is missingness in other positions. In a sense aa_haplotype is more conservative than\n",
    "# ns_changes. We might want to review this decision in future\n",
    "#\n",
    "# We have used '*' to represent multiple unphased synonymous mutations in ns_changes.\n",
    "# This is perhaps undesirable and we might want to review this in the future.\n",
    "#\n",
    "# The current function requires a VCF file. This would probably be much quicker to run if using a zarr representation of the VCF\n",
    "\n",
    "def call_haplotype(\n",
    "#     target_name='crt',\n",
    "    chrom='Pf3D7_07_v3',\n",
    "    starts=[403222, 403490, 403938, 404283, 404569, 404764, 404936, 405146, 405334, 405539, 405825, 406017, 406241], # Start positions of each exon\n",
    "    ends=  [403312, 403758, 404110, 404415, 404640, 404839, 405018, 405196, 405390, 405631, 405869, 406071, 406317], # End positions of each exon\n",
    "    strand='+', # Either '+' or '-'\n",
    "    vcf_file_format=vcf_file_format, # Location of VCFs, allowing chrom to vary\n",
    "    samples=None, # The default (None) will call haplotypes for all samples within the VCF\n",
    "    include_indels=False, # If set to True, this will include indels, and hence potentially output sequences of different lengths\n",
    "    positions_with_missingness=[403620, 403621], # If there are missing genotypes at these positions, these will not be treated as missing. This is important in CRT as many samples have missing genotypes at Pf3D7_07_v3:403620-403621 due to way haplotypes are treated by HaplotypeCaller\n",
    "    positions_to_ignore=[], # These positions will be skipped\n",
    "    positions_to_allow=[403618, 403622], # These allows certain indels to be included even if include_indels is False. Important as common CVIET haplotype in CRT 72-76 is treated by HaplotypeCaller as two 1bp indels, rather than 3 SNPs\n",
    "    ref_fasta_fn=ref_genome_fn,\n",
    "    vcf_fn=None, # If supplied, this will overrule what is provided in vcf_file_format\n",
    "    verbose=False,\n",
    "    show_genotypes=False, # If true, these will give very verbose output\n",
    "    use_filter=True, # If true, will only use PASS variants\n",
    "    first_aa=1, # Important to use if, for example, only creating haplotypes for one exon\n",
    "    to_stop=True # This argument is used by Bio translate function. If True, sequences will be truncated at premature stop codons, resulting in shorter AA sequences\n",
    "):\n",
    "    \n",
    "    from pyfasta import Fasta\n",
    "    from Bio.Seq import Seq\n",
    "    \n",
    "    if not isinstance(starts, list) and not(isinstance(starts, np.ndarray)):\n",
    "        if isinstance(starts, int):\n",
    "            starts = [starts]\n",
    "        else:\n",
    "            sys.exit(\"starts must be a list of integers (or a single integer)\")\n",
    "    if not isinstance(ends, list) and not(isinstance(ends, np.ndarray)):\n",
    "        if isinstance(ends, int):\n",
    "            ends = [ends]\n",
    "        else:\n",
    "            sys.exit(\"ends must be a list of integers (or a single integer)\")\n",
    "    \n",
    "    if len(starts) != len(ends):\n",
    "        sys.exit(\"starts and ends must be the same length\")\n",
    "        \n",
    "    exon_offsets = [0] + list(np.cumsum(np.array(ends)-np.array(starts)+1))[:-1]\n",
    "    \n",
    "    if vcf_fn is None:\n",
    "        vcf_fn = vcf_file_format % chrom\n",
    "    \n",
    "    ref_sequences = []\n",
    "    for i in np.arange(len(starts)):\n",
    "        start = starts[i]\n",
    "        end = ends[i]\n",
    "        ref_sequences.append(Seq(Fasta(ref_fasta_fn)[chrom][starts[i]-1:ends[i]]))\n",
    "    ref_sequence = Seq(\"\")\n",
    "    for s in ref_sequences:\n",
    "        ref_sequence += s\n",
    "    ref_nucleotide_haplotype = str(ref_sequence)\n",
    "    if strand == '+':\n",
    "        ref_aa_haplotype = str(ref_sequence.translate(to_stop=to_stop))\n",
    "    else:\n",
    "        ref_aa_haplotype = str(ref_sequence.reverse_complement().translate(to_stop=to_stop))\n",
    "\n",
    "    vcf_reader = vcf.Reader(filename=vcf_fn)\n",
    "    if samples is None:\n",
    "        samples = vcf_reader.samples\n",
    "    sample_sequences = collections.OrderedDict()\n",
    "    sample_offsets = collections.OrderedDict()\n",
    "    PID = collections.OrderedDict()\n",
    "    non_phased_het_seen = collections.OrderedDict()\n",
    "    sample_unphaseable = collections.OrderedDict()\n",
    "    aa_haplotype = collections.OrderedDict()\n",
    "    nucleotide_haplotype = collections.OrderedDict()\n",
    "    ns_changes = collections.OrderedDict()\n",
    "    num_missing = collections.OrderedDict()\n",
    "    num_called = collections.OrderedDict()\n",
    "    swap_phasing = collections.OrderedDict()\n",
    "    first_PGT = collections.OrderedDict()\n",
    "    for sample in samples:\n",
    "        sample_sequences[sample] = [ref_sequence.tomutable(), ref_sequence.tomutable()]\n",
    "        sample_offsets[sample] = [0, 0]\n",
    "        PID[sample] = ''\n",
    "        non_phased_het_seen[sample] = False\n",
    "        sample_unphaseable[sample] = ''\n",
    "        aa_haplotype[sample] = ''\n",
    "        nucleotide_haplotype[sample] = ''\n",
    "        num_missing[sample] = 0\n",
    "        num_called[sample] = 0\n",
    "        swap_phasing[sample] = False\n",
    "        first_PGT[sample] = ''\n",
    "\n",
    "    for i in np.arange(len(starts)):\n",
    "        start = starts[i]\n",
    "        end = ends[i]\n",
    "        exon_offset = exon_offsets[i]\n",
    "        if verbose:\n",
    "            print(f\"{chrom}:{starts[i]}-{ends[i]}\")\n",
    "        for record in vcf_reader.fetch(chrom, start-1, end): # Find variants in relevant portion of gene\n",
    "            if ( record.POS >= start) and not record.POS in positions_to_ignore: # Have to restrict to position that are at or after start, to ensure indels in previous positions don't mess things up. Note this is a change from previous behaviour\n",
    "                if not use_filter or record.FILTER==[]: # Only use PASS variants, unless use_filter is False\n",
    "                    if verbose:\n",
    "                        print(\"\\n\", record, record.FILTER)\n",
    "                    for sample in samples:\n",
    "                        if show_genotypes:\n",
    "                            print(record.genotype(sample))\n",
    "                        GT = record.genotype(sample)['GT']\n",
    "                        AD = record.genotype(sample)['AD']\n",
    "                        POS = record.POS - start + exon_offset\n",
    "                        if GT == './.' or GT == '.|.': # Note phased empty genotypes (.|.) have appeared in Pf7. These weren't in Pf6!\n",
    "                            if not record.POS in positions_with_missingness:\n",
    "                                num_missing[sample] += 1\n",
    "                        else:\n",
    "                            num_called[sample] += 1\n",
    "                            if GT != '0/0' and GT != '0|0': # Only interested in non-ref genotypes. Hom ref genotypes don't change sequence so can be ignored\n",
    "                                alleles = [record.REF] + record.ALT\n",
    "#                                 first_allele_int = int(GT[0])\n",
    "#                                 second_allele_int = int(GT[2])\n",
    "                                # Find lengths of alleles (to determine which are indels)\n",
    "                                REF_len = len(record.REF)\n",
    "                                first_allele_ALT_len = len(alleles[int(GT[0])])\n",
    "                                second_allele_ALT_len = len(alleles[int(GT[2])])\n",
    "                                if (REF_len==first_allele_ALT_len and REF_len==second_allele_ALT_len) or include_indels or record.POS in positions_to_allow:\n",
    "                                    if (GT[0] != GT[2]) and (alleles[int(GT[0])] != '*') and (alleles[int(GT[2])] != '*'): # heterozygous call at non-spanning deletion\n",
    "                                        if 'PGT' in record.genotype(sample).data._fields:\n",
    "                                            is_unphased = (record.genotype(sample)['PGT'] is None)\n",
    "                                        else:\n",
    "                                            is_unphased = True\n",
    "                                        if non_phased_het_seen[sample]:\n",
    "                                            if verbose:\n",
    "                                                print(\"Sample %s unphased het followed by het\" % sample)\n",
    "                                            sample_unphaseable[sample] = '*'\n",
    "            #                                 genotype[sample] = 'X'\n",
    "                                        if is_unphased:\n",
    "                                            non_phased_het_seen[sample] = True\n",
    "                                        else:\n",
    "#                                             if 'PID' in record.genotype(sample).data._fields: # Note this won't be the case in het spanning deletions\n",
    "                                            if PID[sample] == '':\n",
    "                                                PID[sample] = record.genotype(sample)['PID']\n",
    "                                                first_PGT[sample] = record.genotype(sample)['PGT']\n",
    "                                                if AD[int(GT[2])] > AD[int(GT[0])]:\n",
    "                                                    swap_phasing[sample] = True\n",
    "                                            else:\n",
    "                                                if record.genotype(sample)['PID'] != PID[sample]:\n",
    "                                                    if verbose:\n",
    "                                                        print(\"Sample %s two PIDs\" % sample)\n",
    "                                                    sample_unphaseable[sample] = '*'\n",
    "            #                                         genotype[sample] = '?'\n",
    "#                                             if record.genotype(sample)['PGT'] == '1|0':\n",
    "#                                                 GT = GT[2] + '/' + GT[0]\n",
    "                                        if verbose:\n",
    "                                            print(f\"AD={AD}, GT={GT}\")\n",
    "                                        # Here we use phased genotype (PGT) if in same phasing group as first het\n",
    "                                        if (\n",
    "                                            'PID' in record.genotype(sample).data._fields\n",
    "                                            and ( PID[sample] == record.genotype(sample)['PID'] )\n",
    "                                        ): # In phasing group with first het call\n",
    "                                            if verbose:\n",
    "                                                print(f\"PID: {PID[sample]}, GT: {GT}, PGT: {record.genotype(sample)['PGT']}\")\n",
    "                                            if first_PGT[sample] == record.genotype(sample)['PGT']:\n",
    "                                                if swap_phasing[sample]:\n",
    "                                                    GT = GT[2] + '/' + GT[0]\n",
    "                                            else:\n",
    "                                                if not swap_phasing[sample]:\n",
    "                                                    GT = GT[2] + '/' + GT[0]\n",
    "#                                             GT = record.genotype(sample)['PGT']\n",
    "                                        # Else we ensure we are taking the majority haplotype as the first allele\n",
    "                                        elif AD[int(GT[2])] > AD[int(GT[0])]:\n",
    "                                            GT = GT[2] + '/' + GT[0]\n",
    "                                            if verbose:\n",
    "                                                print(f\"AD={AD}, GT={GT}\")\n",
    "\n",
    "                                for i, sample_offset in enumerate(sample_offsets[sample]):\n",
    "                                    alleles = [record.REF] + record.ALT\n",
    "                                    GTint = int(GT[i*2])\n",
    "                                    REF_len = len(record.REF)\n",
    "                                    ALT_len = len(alleles[GTint])\n",
    "                                    if(verbose):\n",
    "                                        print(f\"before: i={i}, offset={sample_offset}, alleles, GT={GTint}, REF={REF_len}, ALT={ALT_len}, {sample_sequences[sample][i]}\")\n",
    "                                    if GTint != 0 and alleles[GTint] != '*' and (REF_len==ALT_len or include_indels or record.POS in positions_to_allow):\n",
    "                                        sample_sequences[sample][i][POS+sample_offsets[sample][i]:(POS+sample_offsets[sample][i]+REF_len)] = alleles[GTint]\n",
    "                                        sample_offsets[sample][i] = sample_offset + (len(alleles[GTint]) - len(record.REF))\n",
    "                                        if(verbose):\n",
    "                                            print(f\"after : i={i}, offset={sample_offset}, alleles, GT={GTint}, REF={REF_len}, ALT={ALT_len}, {sample_sequences[sample][i]}\")\n",
    "\n",
    "    for sample in samples:\n",
    "        if(show_genotypes):\n",
    "            print(sample_sequences[sample])\n",
    "        ###########################\n",
    "        # populate ns_changes\n",
    "        ###########################\n",
    "        if strand == '+':\n",
    "            aa_0 = (\n",
    "                str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "            aa_1 = (\n",
    "                str(sample_sequences[sample][1].toseq().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "        else:\n",
    "            aa_0 = (\n",
    "                str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "            aa_1 = (\n",
    "                str(sample_sequences[sample][1].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "#         if aa_0 != '!' and aa_1 != '!' # This was what was used in Pf6\n",
    "        if aa_0 != '!' and aa_1 != '!' and (len(aa_0) == len(ref_aa_haplotype)) and (len(aa_1) == len(ref_aa_haplotype)): # In Pf7, we have some in-frame indels, hence we also need to check for these, and give \"!\" if indels exist\n",
    "            if aa_0 == aa_1:\n",
    "                ns_changes_list = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "                    if aa_0[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_0[i]))\n",
    "                ns_changes[sample] = \"/\".join(ns_changes_list)\n",
    "            else:\n",
    "                ns_changes_list_0 = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "                    if aa_0[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list_0.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_0[i]))\n",
    "                ns_changes_0 = \"/\".join(ns_changes_list_0)\n",
    "                ns_changes_list_1 = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "#                     print(len(ref_aa_haplotype), len(aa_1))\n",
    "                    if aa_1[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list_1.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_1[i]))\n",
    "                ns_changes_1 = \"/\".join(ns_changes_list_1)\n",
    "\n",
    "                if ns_changes_0 == '':\n",
    "                    ns_changes[sample] = ns_changes_1.lower()\n",
    "                elif ns_changes_1 == '':\n",
    "                    ns_changes[sample] = ns_changes_0.lower()\n",
    "                else:\n",
    "                    ns_changes[sample] = (\",\".join([ns_changes_0, ns_changes_1])).lower()\n",
    "        else:\n",
    "            if verbose:\n",
    "                print(aa_0, aa_1, ref_aa_haplotype)\n",
    "            ns_changes[sample] = '!'\n",
    "\n",
    "        ################################################\n",
    "        # populate nucleotide_haplotype and aa_haplotype\n",
    "        ################################################\n",
    "        if str(sample_sequences[sample][0]).upper() == str(sample_sequences[sample][1]).upper():\n",
    "            nucleotide_haplotype[sample] = str(sample_sequences[sample][0].toseq())\n",
    "            if strand == '+':\n",
    "                aa_haplotype[sample] = ( str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                                    if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                                    else '!' )\n",
    "            else:\n",
    "                aa_haplotype[sample] = ( str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                                    if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                                    else '!' )\n",
    "        else:\n",
    "            nucleotide_haplotype[sample] = \"%s,%s\" % ( str(sample_sequences[sample][0].toseq()), str(sample_sequences[sample][1].toseq()) )\n",
    "            if strand == '+':\n",
    "                aa_haplotype[sample] = \"%s,%s\" % (\n",
    "                    ( str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                     else '!' ),\n",
    "                    ( str(sample_sequences[sample][1].toseq().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                     else '!' )\n",
    "                )\n",
    "            else:\n",
    "                aa_haplotype[sample] = \"%s,%s\" % (\n",
    "                    ( str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                     else '!' ),\n",
    "                    ( str(sample_sequences[sample][1].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                     else '!' )\n",
    "                )\n",
    "    \n",
    "        # Add asterisk to end if sample is unphaseable\n",
    "        ns_changes[sample] += sample_unphaseable[sample]\n",
    "        aa_haplotype[sample] += sample_unphaseable[sample]\n",
    "        nucleotide_haplotype[sample] += sample_unphaseable[sample]\n",
    "\n",
    "        # Set ns_changes output to missing ('-') if no NS changes and at least one missing genotype call in the region\n",
    "        if (ns_changes[sample] == '' or ns_changes[sample] == '*') and num_missing[sample] > 0:\n",
    "            ns_changes[sample] = '-'\n",
    "                \n",
    "        # Set sequence output to missing ('-') if any missing genotype call in the gene\n",
    "        if num_missing[sample] > 0:\n",
    "            aa_haplotype[sample] = '-'\n",
    "            nucleotide_haplotype[sample] = '-'\n",
    "    \n",
    "    df_all_haplotypes = pd.DataFrame(\n",
    "        {\n",
    "            'aa_haplotype': pd.Series(aa_haplotype),\n",
    "            'nucleotide_haplotype': pd.Series(nucleotide_haplotype),\n",
    "            'ns_changes': pd.Series(ns_changes),\n",
    "        }\n",
    "    )\n",
    "        \n",
    "        \n",
    "#     return(df_haplotypes, ref_aa_haplotype, ref_nucleotide_haplotype)\n",
    "    return(\n",
    "        {\n",
    "            'df_all_haplotypes':df_all_haplotypes,\n",
    "            'ref_aa_haplotype':ref_aa_haplotype,\n",
    "            'ref_nucleotide_haplotype':ref_nucleotide_haplotype,\n",
    "        }\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_plot_data(\n",
    "    gene_id='PF3D7_0709000',\n",
    "    haplotype_calls=None,\n",
    "    samples_fn=samples_fn,\n",
    "    expected_aa_len=None,\n",
    "#     populations = ['SA', 'AF-W', 'AF-C', 'AF-NE', 'AF-E', 'AS-S-E', 'AS-S-FE', 'AS-SE-W', 'AS-SE-E', 'OC-NG', ],\n",
    "    populations = ['AF-W', 'AF-C', 'AF-NE', 'AF-E', ],\n",
    "    samples_to_show = {\n",
    "        'PG0563-C': '7G8',\n",
    "        'PG0564-C': 'GB4',\n",
    "        'PG0565-C': 'HB3',\n",
    "        'PG0566-C': 'Dd2',\n",
    "        'PG0567-C': '3D7',\n",
    "        'PG0568-C': 'IT',\n",
    "    },\n",
    "    sample_to_use_as_background = 'PG0566-C', # Dd2\n",
    "    verbose=True,\n",
    "    **kwargs\n",
    "):\n",
    "#     print(gene_id)\n",
    "    if gene_id is not None:\n",
    "        cds_results = determine_cds(gene_id)\n",
    "#     print(cds_results['gene_name'])\n",
    "    \n",
    "        if haplotype_calls is None:\n",
    "            haplotype_calls = call_haplotype(\n",
    "                chrom=cds_results['chrom'],\n",
    "                starts=cds_results['starts'],\n",
    "                ends=cds_results['ends'],\n",
    "                strand=cds_results['strand'],\n",
    "                **kwargs\n",
    "            )\n",
    "    else:\n",
    "        haplotype_calls = call_haplotype(\n",
    "            **kwargs\n",
    "        )\n",
    "        \n",
    "    \n",
    "    df_samples = pd.read_csv(samples_fn, sep='\\t', low_memory=False, index_col=0)\n",
    "    if verbose:\n",
    "        print(f\"Read in {df_samples.shape[0]:,} samples, of which {np.count_nonzero(df_samples['QC pass']):,} passed QC\")\n",
    "\n",
    "    df_join = df_samples.join(haplotype_calls['df_all_haplotypes'])\n",
    "    \n",
    "    if expected_aa_len is None:\n",
    "        expected_aa_len = len(haplotype_calls['ref_aa_haplotype'])\n",
    "    \n",
    "    df_clonal = df_join.loc[\n",
    "        (\n",
    "            ( df_join['QC pass'] == True )\n",
    "            & ( df_join['aa_haplotype'].astype(str).apply(len) == expected_aa_len )\n",
    "        ) | ( df_join['Study'] == '1153-PF-Pf3KLAB-KWIATKOWSKI' ),\n",
    "        [\n",
    "            'aa_haplotype',\n",
    "            'nucleotide_haplotype',\n",
    "            'ns_changes',\n",
    "            'Population',\n",
    "        ]\n",
    "    ]\n",
    "    \n",
    "    if df_clonal.shape[0] == 0:\n",
    "        most_common_length = df_join['aa_haplotype'].astype(str).apply(len).value_counts().index[0]\n",
    "        raise Exception(f\"There were no samples with AA haplotype length = {expected_aa_len}. Most common length was {most_common_length}\")\n",
    "    \n",
    "    df_clonal['ns_changes'].fillna('', inplace=True)\n",
    "    df_clonal['number_of_mutations'] = df_clonal['ns_changes'].apply(lambda x: len(x.split('/')))\n",
    "    df_clonal.loc[df_clonal['ns_changes'] == '', 'number_of_mutations'] = 0\n",
    "    if verbose:\n",
    "        print(f\"Found {df_clonal.shape[0]:,} clonal haplotypes in QC pass samples\")\n",
    "\n",
    "    grouping_columns = [\n",
    "        'number_of_mutations', 'ns_changes',\n",
    "    #     'haplotype_name',\n",
    "    ]\n",
    "    ascending=[True]*len(grouping_columns)\n",
    "    ascending.append(False)\n",
    "\n",
    "    def n_agg(x):\n",
    "        names = collections.OrderedDict()\n",
    "        for population in populations:\n",
    "            names[population] = np.count_nonzero( (x['Population'] == population) )\n",
    "        return pd.Series(names)\n",
    "\n",
    "    df_haplotypes = df_clonal.groupby(grouping_columns).apply(n_agg)\n",
    "    df_haplotypes['Total'] = df_haplotypes.sum(axis=1)\n",
    "    df_haplotypes.reset_index(inplace=True)\n",
    "    df_haplotypes.sort_values(grouping_columns + ['Total'], ascending=ascending, inplace=True)\n",
    "    df_haplotypes['ns_changes_list'] = df_haplotypes['ns_changes'].apply(lambda x: x.split('/'))\n",
    "    df_haplotypes.sort_values('Total', ascending=False, inplace=True)\n",
    "    # Determine which samples to show have which haplotypes\n",
    "    df_haplotypes['sample_names'] = ''\n",
    "    for sample in samples_to_show.keys():\n",
    "        ns_changes = df_clonal.loc[sample, 'ns_changes']\n",
    "        if df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'].values == '':\n",
    "            df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] = samples_to_show[sample]\n",
    "        else:\n",
    "            df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] = (\n",
    "                df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] + '\\n' + samples_to_show[sample]\n",
    "            )\n",
    "    if verbose:\n",
    "        print(f\"Found {df_haplotypes.shape[0]:,} unique haplotypes in QC pass samples\")\n",
    "        \n",
    "    # Determine background mutations\n",
    "    if sample_to_use_as_background in df_clonal.index:\n",
    "        background_ns_changes = df_clonal.loc[sample_to_use_as_background, 'ns_changes']\n",
    "        \n",
    "    if gene_id is not None:\n",
    "        gene_name = cds_results['gene_name']\n",
    "    else:\n",
    "        gene_name = None\n",
    "    \n",
    "    \n",
    "    return(\n",
    "        {\n",
    "            'df_haplotypes':df_haplotypes,\n",
    "            'df_clonal':df_clonal,\n",
    "            'df_join':df_join,\n",
    "            'background_ns_changes':background_ns_changes,\n",
    "            'haplotype_calls':haplotype_calls,\n",
    "            'gene_name':gene_name,\n",
    "            'df_samples':df_samples,\n",
    "        }\n",
    "    )\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot for DRF\n",
    "\n",
    "def drug_plot(drug, drug_name, populations, label): \n",
    "    \n",
    "    from matplotlib.ticker import MaxNLocator\n",
    "\n",
    "    plt.rcParams['axes.labelsize'] = 4\n",
    "    plt.rcParams['axes.titlesize'] = 5\n",
    "    \n",
    "    plt.rcParams['xtick.labelsize'] = 3\n",
    "    plt.rcParams['ytick.labelsize'] = 4\n",
    "    \n",
    "    plt.rcParams['ytick.alignment'] = 'bottom'\n",
    "    \n",
    "    plt.rcParams['figure.dpi'] = 1000\n",
    "    \n",
    "    \n",
    "    xlims = [(1982, 1986),\n",
    "            (1994,1998),\n",
    "            (2000,2019)]\n",
    "    \n",
    "    fig, axs = plt.subplots(1, 3, gridspec_kw={'width_ratios': [1,1,5]}, figsize=((16/5),9), sharey=True)\n",
    "                \n",
    "    ax, ax1, ax2 = axs\n",
    "    fig.subplots_adjust(hspace=0.05)\n",
    "    \n",
    "    first_label_ax2 = True \n",
    "    \n",
    "    for i,ax in enumerate(axs):\n",
    "\n",
    "        for pop in reversed(populations):\n",
    "            \n",
    "            country = drug[drug['Population'] == pop]['Country']\n",
    "            _, idx = np.unique(country, return_index=True)\n",
    "            country = country.reset_index(drop=True)\n",
    "\n",
    "            for c in country[np.sort(idx)]:\n",
    "\n",
    "                data = drug[(drug['Population'] == pop) & (drug['Country'] == c)]\n",
    "\n",
    "                for index, row in data.iterrows():\n",
    "                    \n",
    "                    if  (row[drug_name + ' frequency'] == 1):\n",
    "                        ax.plot('Year', 'Label', '-o', data=row, color = 'black',markeredgewidth=0.3, markersize=4)\n",
    "                        # plot x marker for one DR frequency \n",
    "                        ax.plot('Year', 'Label', '-x', data=row, color = 'white',markeredgewidth=0.3, markersize=4)\n",
    "                        \n",
    "                    elif  (row[drug_name + ' frequency'] == 0):\n",
    "                        ax.plot('Year', 'Label', '-o', data=row, markeredgecolor= 'black',  markerfacecolor='None',  markeredgewidth=0.3, markersize=4)\n",
    "                        # plot x marker for zero DR frequency \n",
    "                        ax.plot('Year', 'Label', '-x', data=row, color = 'black', markeredgewidth=0.3, markersize=4)\n",
    "                    \n",
    "    \n",
    "                    elif np.isnan(row[drug_name + ' frequency']) :\n",
    "\n",
    "                        # add the  black edge colour for each point\n",
    "                        ax.plot('Year', 'Label', '-o', data=row, markeredgecolor= 'white', alpha=0,  markerfacecolor='None', markersize=4)\n",
    "\n",
    "                    else:\n",
    "                        # add the  black edge colour for each point\n",
    "                        ax.plot('Year', 'Label', '-o', data=row, markeredgecolor= 'black',  markerfacecolor='None',  markeredgewidth=0.3, markersize=4)\n",
    "                        # plot DR frequency with alpha value\n",
    "                        ax.plot('Year', 'Label', '-o', data=row, color = 'black', alpha=(row[drug_name +' frequency']),  markeredgewidth=0.3, markersize=4)\n",
    "\n",
    "                    x1,x2 = xlims[i]\n",
    "                    ax.set_xlim(x1,x2)\n",
    "                    \n",
    "                    ax.grid(axis='y', color = 'grey', linewidth = '0.25')\n",
    "                        \n",
    "                    if i == 2:\n",
    "                        ax.spines['left'].set_visible(False)\n",
    "                        ax.tick_params(axis='y', colors='None')\n",
    "                        #ax.set_axisbelow(True)\n",
    "                       \n",
    "                        \n",
    "                    elif i == 1:\n",
    "                        ax.spines['right'].set_visible(False)\n",
    "                        ax.spines['left'].set_visible(False)\n",
    "                        #ax.set_axisbelow(True)\n",
    "                        ax.tick_params(axis='y', colors='None')\n",
    "                        \n",
    "\n",
    "                        \n",
    "                    \n",
    "                    elif i == 0:\n",
    "\n",
    "                        ax.spines['right'].set_visible(False)\n",
    "                        ax.tick_params(axis='y', colors='None')\n",
    "                        \n",
    "                    ax.xaxis.set_minor_locator(plt.MaxNLocator(x2-x1))\n",
    "                    \n",
    "                            \n",
    "\n",
    "    ax.set_title(drug_name, loc='left')\n",
    "    \n",
    "    #x axis years as integers           \n",
    "    ax.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    \n",
    "    #set yaxis with population colours\n",
    "    labels = []\n",
    "#     labels += [population_colours['OC-NG']]*label['OC-NG']\n",
    "#     labels += [population_colours['AS-SE-E']]*label['AS-SE-E']\n",
    "#     labels += [population_colours['AS-SE-W']]*label['AS-SE-W']\n",
    "\n",
    "#     labels += [population_colours['AS-S-FE']]*label['AS-S-FE']\n",
    "#     labels += [population_colours['AS-S-E']]*label['AS-S-E']\n",
    "\n",
    "    labels += [population_colours['AF-E']]*label['AF-E']\n",
    "    labels += [population_colours['AF-NE']]*label['AF-NE']\n",
    "    labels += [population_colours['AF-C']]*label['AF-C']\n",
    "    labels += [population_colours['AF-W']]*label['AF-W']\n",
    "#     labels += [population_colours['SA']]*label['SA']\n",
    "\n",
    "\n",
    "    ax.tick_params(axis='y',direction='out', pad=58, reset = True)\n",
    "   \n",
    "\n",
    "    [(t.set_color(labels[n]),t.set_horizontalalignment('left'), t.set_fontsize(4)) for n,t in enumerate(ax.yaxis.get_ticklabels())]\n",
    "\n",
    "    ax.tick_params(axis='y', which='both', length=0, bottom = True)\n",
    "    ax.grid(axis='y', color = 'grey', linewidth = '0.25')\n",
    "                        \n",
    "                        \n",
    "    plt.subplots_adjust(wspace=0.15)\n",
    "    \n",
    "    return(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locations_agg(x, ns_changes='F313S/Q454E'):\n",
    "    names = collections.OrderedDict()\n",
    "    names['n'] = np.count_nonzero(x['ns_changes_homozygous'])\n",
    "    if names['n'] == 0:\n",
    "        names[f'{ns_changes} frequency'] = np.nan\n",
    "    else:\n",
    "        names[f'{ns_changes} frequency'] = np.count_nonzero(\n",
    "            ( x['ns_changes'] == ns_changes)\n",
    "        ) / names['n']\n",
    "    \n",
    "    return pd.Series(names)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def haplotype_abacus_plot(\n",
    "    ns_changes = 'F313S/Q454E',\n",
    "    min_samples = 25,\n",
    "    ppd_results = None,\n",
    "    df_samples = None,\n",
    "#     populations = ['SA', 'AF-W', 'AF-C', 'AF-NE', 'AF-E', 'AS-S-E', 'AS-S-FE', 'AS-SE-W', 'AS-SE-E', 'OC-NG'],\n",
    "    populations = ['AF-W', 'AF-C', 'AF-NE', 'AF-E', ],\n",
    "    png_fn=None,\n",
    "    dpi=240,\n",
    "    pdf_fn=None,\n",
    "):\n",
    "    df_samples_with_ns_changes = df_samples.join(ppd_results['haplotype_calls']['df_all_haplotypes']['ns_changes'])\n",
    "    df_samples_with_ns_changes = df_samples_with_ns_changes.loc[df_samples_with_ns_changes['QC pass']]\n",
    "    df_samples_with_ns_changes.loc[df_samples_with_ns_changes['Country'] == 'Democratic Republic of the Congo', ['Country']] = 'DRC'\n",
    "\n",
    "    df_samples_with_ns_changes['ns_changes_homozygous'] = ( df_samples_with_ns_changes['ns_changes'] == df_samples_with_ns_changes['ns_changes'].str.upper() )\n",
    "    \n",
    "    df_frequencies = (\n",
    "        df_samples_with_ns_changes\n",
    "        .groupby(['Population', 'Year', 'Country', 'Admin level 1'])\n",
    "        .apply(lambda x: locations_agg(x, ns_changes))\n",
    "        .reset_index()\n",
    "        .set_index(['Population', 'Country', 'Admin level 1'])\n",
    "        .loc[locations]\n",
    "        .reset_index()\n",
    "    )\n",
    "    df_frequencies = df_frequencies.loc[df_frequencies['n'] >= min_samples]\n",
    "    df_frequencies['Label'] = df_frequencies['Country'] + ', ' + df_frequencies['Admin level 1']\n",
    "    \n",
    "    label = {}\n",
    "    for p in populations: \n",
    "        arr = np.unique(df_frequencies[df_frequencies['Population'] == p]['Admin level 1'], return_counts= True)\n",
    "        label.update({p: len(arr[1])})\n",
    "\n",
    "    fig = drug_plot(df_frequencies, ns_changes, populations, label)\n",
    "\n",
    "    if png_fn is not None:\n",
    "        fig.savefig(png_fn, dpi=dpi)\n",
    "    if pdf_fn is not None:\n",
    "        fig.savefig(pdf_fn)\n",
    "\n",
    "#     return(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def haplotypes_plot(\n",
    "    gene_id=None,\n",
    "    prepare_plot_data_results=None,\n",
    "    gene_name='',\n",
    "    min_samples=5,\n",
    "    figsize=(12, 12),\n",
    "    max_ylim=5000,\n",
    "    sample_counts_plot_log='y',\n",
    "    cumulative_proportions_text=None,\n",
    "    background_ns_changes=None,\n",
    "    create_abacus_plots=True,\n",
    "    min_samples_for_abacus_plots=25,\n",
    "    png_fn=None,\n",
    "    dpi=240,\n",
    "    pdf_fn=None,\n",
    "#     background_ns_changes='M74I/N75E/K76T/A220S/Q271E/N326S/I356T/R371I',\n",
    "    **kwargs\n",
    "):\n",
    "    if cumulative_proportions_text is None:\n",
    "        if sample_counts_plot_log == 'y':\n",
    "            cumulative_proportions_text = 'below'\n",
    "        else:\n",
    "            cumulative_proportions_text = 'above'\n",
    "    \n",
    "    return_values = 'fig'\n",
    "    \n",
    "    if prepare_plot_data_results is None:\n",
    "        return_values = 'all'\n",
    "        prepare_plot_data_results = prepare_plot_data(gene_id=gene_id, **kwargs)\n",
    "    df_haplotypes = prepare_plot_data_results['df_haplotypes']\n",
    "    if gene_name == '':\n",
    "        gene_name = prepare_plot_data_results['gene_name']\n",
    "\n",
    "    if background_ns_changes is None:\n",
    "        background_ns_changes = prepare_plot_data_results['background_ns_changes']\n",
    "\n",
    "    fig, axs = plt.subplots(3, 1, figsize=figsize, gridspec_kw={'height_ratios': [1, 0.5, 2]})\n",
    "    \n",
    "    total_samples = df_haplotypes['Total'].sum()\n",
    "    df_haplotypes['cum_proportion'] = df_haplotypes['Total'].cumsum() / total_samples\n",
    "    \n",
    "    df_haplotypes_set = df_haplotypes.loc[df_haplotypes['Total'] >= min_samples]\n",
    "\n",
    "    mutations = pd.Series(np.unique(np.concatenate(df_haplotypes_set['ns_changes_list'].values)))\n",
    "    mutations = mutations.loc[mutations != '']\n",
    "    mutations_np_ref = mutations.apply(lambda x: x[1:])\n",
    "    aa = mutations_np_ref.apply(lambda x: int(x[:-1]))\n",
    "    df_mutations_set = pd.DataFrame(\n",
    "        {\n",
    "            'mutation': mutations.values,\n",
    "            'aa': aa.values,\n",
    "        }\n",
    "    ).sort_values('aa').reset_index(drop=True).reset_index().set_index('mutation')\n",
    "\n",
    "    # Breakdown by number of samples (log scale)\n",
    "    axs[0].bar(df_haplotypes_set['ns_changes'], df_haplotypes_set['Total'], color='black', log=sample_counts_plot_log)\n",
    "    axs[0].set_xticklabels([])\n",
    "    if sample_counts_plot_log == 'y':\n",
    "        axs[0].set_ylim(0.5, max_ylim)\n",
    "    if sample_counts_plot_log=='y':\n",
    "        yticklabel_values = [0, 0] + [10**x for x in np.arange(int(np.ceil(np.log10(max_ylim + 1))))]\n",
    "        axs[0].set_yticklabels(yticklabel_values)\n",
    "        axs[0].set_ylabel('Number of samples (log scale)', fontsize=10)\n",
    "    else:\n",
    "        axs[0].set_ylabel('Number of samples', fontsize=10)\n",
    "    if cumulative_proportions_text=='below':\n",
    "        for i in np.arange(df_haplotypes_set.shape[0]):\n",
    "            axs[0].text(\n",
    "                i,\n",
    "                df_haplotypes_set.iloc[i]['Total'],\n",
    "                f\"{df_haplotypes_set.iloc[i]['cum_proportion']:.1%}\",\n",
    "                rotation=90,\n",
    "                ha='center',\n",
    "                va='top',\n",
    "                color='white'\n",
    "            )\n",
    "    if cumulative_proportions_text=='above':\n",
    "        for i in np.arange(df_haplotypes_set.shape[0]):\n",
    "            axs[0].text(\n",
    "                i,\n",
    "                df_haplotypes_set.iloc[i]['Total'],\n",
    "                f\"{df_haplotypes_set.iloc[i]['cum_proportion']:.1%}\",\n",
    "                rotation=90,\n",
    "                ha='center',\n",
    "                va='bottom',\n",
    "                color='black'\n",
    "            )\n",
    "\n",
    "    # Breakdown by population\n",
    "    for column_num in np.arange(2, df_haplotypes_set.shape[1]-4):\n",
    "        pop = df_haplotypes_set.columns[column_num]\n",
    "#         print(pop)\n",
    "\n",
    "        label=pop\n",
    "        if column_num == 2:\n",
    "            axs[1].bar(\n",
    "                df_haplotypes_set['ns_changes'],\n",
    "                df_haplotypes_set.iloc[:, column_num] / df_haplotypes_set.loc[:, 'Total'],\n",
    "                color=population_colours[pop],\n",
    "                label=label\n",
    "            )\n",
    "        else:\n",
    "            axs[1].bar(\n",
    "                df_haplotypes_set['ns_changes'],\n",
    "                df_haplotypes_set.iloc[:, column_num] / df_haplotypes_set.loc[:, 'Total'],\n",
    "                color=population_colours[pop],\n",
    "                label=label, \n",
    "                bottom=df_haplotypes_set.iloc[:, 2:column_num].sum(axis=1) / df_haplotypes_set.loc[:, 'Total'],\n",
    "            )\n",
    "\n",
    "    axs[1].set_xticklabels(df_haplotypes_set['sample_names'])\n",
    "    \n",
    "    l = axs[1].legend(loc=\"lower left\",  ncol = len(population_colours), bbox_to_anchor=(0.1, 1.0), frameon=False)\n",
    "\n",
    "    # Mutations in each haplotype\n",
    "    i = 0\n",
    "    if ( not '' in background_ns_changes ) and background_ns_changes in df_haplotypes_set['ns_changes'].values:\n",
    "        background_mutation_indices = df_mutations_set.loc[\n",
    "            df_haplotypes_set.loc[\n",
    "                df_haplotypes_set['ns_changes'] == background_ns_changes,\n",
    "                'ns_changes_list'\n",
    "            ].values[0],\n",
    "            'index'\n",
    "        ].values\n",
    "    else:\n",
    "        background_mutation_indices = []\n",
    "\n",
    "    for ix, row in df_haplotypes_set.iterrows():\n",
    "        if row['ns_changes_list'][0] != '':\n",
    "            indexes = df_mutations_set.loc[row['ns_changes_list'], 'index'].values\n",
    "            axs[2].plot([i] * len(indexes), indexes, '-', linewidth=3, color='grey')\n",
    "            background_mutations = np.intersect1d(indexes, background_mutation_indices)\n",
    "            other_mutations = np.setdiff1d(indexes, background_mutation_indices)\n",
    "            axs[2].scatter([i] * len(background_mutations), background_mutations, s=100, color='grey', zorder=3)\n",
    "            axs[2].scatter([i] * len(other_mutations), other_mutations, s=100, color='black', zorder=3)\n",
    "        i = i + 1\n",
    "    axs[2].set_xlim(axs[0].get_xlim())\n",
    "    axs[2].set_xticks([])\n",
    "    axs[2].set_yticks(np.arange(df_mutations_set.shape[0]))\n",
    "    axs[2].set_yticklabels(df_mutations_set.index)\n",
    "    axs[2].grid(axis = 'y')\n",
    "    axs[2].spines['top'].set_visible(False)\n",
    "    axs[2].spines['right'].set_visible(False)\n",
    "    axs[2].spines['bottom'].set_visible(False)\n",
    "    axs[2].spines['left'].set_visible(False)\n",
    "\n",
    "    axs[1].set_ylabel('Proportion of samples', fontsize=10)\n",
    "#     axs[2].set_xlabel('All $crt$ haplotypes with Dd2 background', fontsize=10)\n",
    "    axs[2].set_xlabel('Haplotypes', fontsize=10)\n",
    "\n",
    "    axs[0].set_title(gene_name)\n",
    "    \n",
    "    fig.tight_layout()\n",
    "    \n",
    "    if png_fn is not None:\n",
    "        fig.savefig(png_fn, dpi=dpi)\n",
    "    if pdf_fn is not None:\n",
    "        fig.savefig(pdf_fn)\n",
    "        \n",
    "    \n",
    "    if create_abacus_plots:\n",
    "        for ns_changes in df_haplotypes_set['ns_changes']:\n",
    "#             print(ns_changes)\n",
    "            if png_fn is None:\n",
    "                abacus_png_fn = None\n",
    "            else:\n",
    "                abacus_png_fn = png_fn.replace(\".png\", f\"_{ns_changes.replace('/', '_')}.png\")\n",
    "            if pdf_fn is None:\n",
    "                abacus_pdf_fn = None\n",
    "            else:\n",
    "                abacus_pdf_fn = pdf_fn.replace(\".pdf\", f\"_{ns_changes.replace('/', '_')}.pdf\")\n",
    "            haplotype_abacus_plot(\n",
    "                ns_changes = ns_changes,\n",
    "                min_samples = min_samples_for_abacus_plots,\n",
    "                ppd_results = prepare_plot_data_results,\n",
    "                df_samples = prepare_plot_data_results['df_samples'],\n",
    "                png_fn = abacus_png_fn,\n",
    "                pdf_fn = abacus_pdf_fn,\n",
    "            )\n",
    "    \n",
    "    if return_values == 'all':\n",
    "        return(\n",
    "            {\n",
    "                'fig':fig,\n",
    "                'df_haplotypes':prepare_plot_data_results['df_haplotypes'],\n",
    "                'df_clonal':prepare_plot_data_results['df_clonal'],\n",
    "                'df_join':prepare_plot_data_results['df_join'],\n",
    "                'background_ns_changes':background_ns_changes,\n",
    "                'haplotype_calls':prepare_plot_data_results['haplotype_calls'],\n",
    "            }\n",
    "        )\n",
    "    elif return_values == 'fig':\n",
    "        return(fig)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_of_interest = collections.OrderedDict()\n",
    "\n",
    "genes_of_interest['PF3D7_0417200'] = 'DHFR'\n",
    "genes_of_interest['PF3D7_0523000'] = 'MDR1'\n",
    "genes_of_interest['PF3D7_0709000'] = 'CRT'\n",
    "genes_of_interest['PF3D7_0810800'] = 'DHPS'\n",
    "genes_of_interest['PF3D7_1224000'] = 'GCH1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: agg\n",
      "PF3D7_0417200 DHFR\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 13,781 clonal haplotypes in QC pass samples\n",
      "Found 30 unique haplotypes in QC pass samples\n",
      "7\n",
      "AF-W\n",
      "AF-C\n",
      "AF-NE\n",
      "AF-E\n",
      "\n",
      "PF3D7_0523000 MDR1\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 10,239 clonal haplotypes in QC pass samples\n",
      "Found 323 unique haplotypes in QC pass samples\n",
      "7\n",
      "AF-W\n",
      "AF-C\n",
      "AF-NE\n",
      "AF-E\n",
      "\n",
      "PF3D7_0709000 CRT\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 12,405 clonal haplotypes in QC pass samples\n",
      "Found 101 unique haplotypes in QC pass samples\n",
      "7\n",
      "AF-W\n",
      "AF-C\n",
      "AF-NE\n",
      "AF-E\n",
      "\n",
      "PF3D7_0810800 DHPS\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 12,533 clonal haplotypes in QC pass samples\n",
      "Found 78 unique haplotypes in QC pass samples\n",
      "7\n",
      "AF-W\n",
      "AF-C\n",
      "AF-NE\n",
      "AF-E\n",
      "\n",
      "PF3D7_1224000 GCH1\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 14,886 clonal haplotypes in QC pass samples\n",
      "Found 97 unique haplotypes in QC pass samples\n",
      "7\n",
      "AF-W\n",
      "AF-C\n",
      "AF-NE\n",
      "AF-E\n",
      "\n",
      "CPU times: user 30min 32s, sys: 42.3 s, total: 31min 14s\n",
      "Wall time: 31min 44s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "%matplotlib\n",
    "# %matplotlib inline\n",
    "results=collections.OrderedDict()\n",
    "for gene_id in genes_of_interest:\n",
    "    print(gene_id, genes_of_interest[gene_id])\n",
    "    !mkdir -p plots/png/{gene_id}\n",
    "    png_fn = f\"plots/png/{gene_id}/{gene_id}.png\"\n",
    "    !mkdir -p plots/pdf/{gene_id}\n",
    "    pdf_fn = f\"plots/pdf/{gene_id}/{gene_id}.pdf\"\n",
    "    results[genes_of_interest[gene_id]] = haplotypes_plot(gene_id=gene_id, sample_counts_plot_log='', png_fn=png_fn, pdf_fn=pdf_fn)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 21 20:05:54 GMT 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusions\n",
    "\n",
    "Here we have created haplotype and \"abacus\" plots for the five genes of interest. Some brief highlights from a quick look at the output plots include:\n",
    "- There are two haplotypes containing DHPS A581G\n",
    "    - K540E/A581G is seen mostly in East Africa, though very rarely in Kenya\n",
    "    - I431V/S436A/A581G/A613S is seen exclusively in west and central Africa, and is most common in Benin, Nigeria and Cameroon\n",
    "- We see two different double-mutants of DHFR, C59R/S108N and N51I/S108N, with the latter being relatively rare in west Africa, but almost at fixation in Sudan. \n",
    "- We don't see more than one amino acid mutation in any haplotype of GCH1 seen in at least 5 samples.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
