{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Oct 30 14:54:29 GMT 2023\n",
      "gen3-os0000012\n",
      "/nfs/team112_data03/personal/nw20/gitlab/malariagen/gsp/mutation-discovery-app/work/produce_gene_summaries\n"
     ]
    }
   ],
   "source": [
    "!date\n",
    "!hostname\n",
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test imports list\n",
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import vcf\n",
    "import collections\n",
    "from Bio.Seq import Seq\n",
    "import allel\n",
    "from pyfasta import Fasta\n",
    "import pickle\n",
    "#import matplotlib.pyplot as plt\n",
    "#from matplotlib.ticker import MaxNLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are loading:\n",
    "#  a vcf (eventually zarr - this can be pf7_variants)\n",
    "#  metadata\n",
    "#  ref genome\n",
    "#  Genome annotations\n",
    "# Inputs\n",
    "vcf_file_format = \"/lustre/scratch124/gsu/legacy/pipelines/builds/pf_70_build/pf_70_internal_release/vcf/%s.pf7.vcf.gz\" # Location of VCFs, allowing chrom to vary\n",
    "samples_fn = '/lustre/scratch126/gsu/team112/pf7/ftp_20221021/Pf7_samples.txt'\n",
    "ref_genome_fn = '/lustre/scratch124/gsu/legacy/pfalciparum/resources/Pfalciparum.genome.fasta'\n",
    "gff_fn = '/lustre/scratch124/gsu/legacy/pfalciparum/resources/snpEff/data/Pfalciparum_GeneDB_Feb2020/Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene_name': 'CRT',\n",
       " 'chrom': 'Pf3D7_07_v3',\n",
       " 'strand': '+',\n",
       " 'starts': array([403222, 403490, 403938, 404283, 404569, 404764, 404936, 405146,\n",
       "        405334, 405539, 405825, 406017, 406241]),\n",
       " 'ends': array([403312, 403758, 404110, 404415, 404640, 404839, 405018, 405196,\n",
       "        405390, 405631, 405869, 406071, 406317])}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def determine_cds(gene_id='PF3D7_0709000', gff_fn=gff_fn):\n",
    "    df_gff = allel.gff3_to_dataframe(gff_fn, attributes=['ID', 'Name'])\n",
    "    gene_name = df_gff.loc[df_gff['ID'] == gene_id, 'Name'].values[0]\n",
    "    if (gene_name == '') or (gene_name == '.'):\n",
    "        gene_name=gene_id\n",
    "    chrom = df_gff.loc[df_gff['ID'] == gene_id, 'seqid'].values[0]\n",
    "    strand = df_gff.loc[df_gff['ID'] == gene_id, 'strand'].values[0]\n",
    "    df_cds = df_gff.loc[\n",
    "        ( df_gff['ID'].str.startswith(gene_id) )\n",
    "        & ( df_gff['type'] == 'CDS' )\n",
    "    ]\n",
    "    return(\n",
    "        {\n",
    "            'gene_name':gene_name,\n",
    "            'chrom':chrom,\n",
    "            'strand':strand,\n",
    "            'starts':df_cds['start'].values,\n",
    "            'ends':df_cds['end'].values,\n",
    "        }\n",
    "    )\n",
    "\n",
    "determine_cds()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# call_haplotype function\n",
    "####################################\n",
    "\n",
    "# This function creates nucleotide and amino acid haplotypes of specific regions, for example:\n",
    "# 1) A single amino acid such as CRT 76\n",
    "# 2) A stretch of contiguous amino acids such as CRT 72-76\n",
    "# 3) The full coding region of a gene such as CRT\n",
    "\n",
    "# Note that in the case of 3), if the gene has multiple exons, the region of interest will be non-contiguous\n",
    "\n",
    "# The function can be used to call a) a single sample, b) specific samples of interest or c) all samples in the VCF (the default).\n",
    "# Note that calling all samples isn't much slower than calling a single sample. Calling all haplotypes for all Pf7 samples took ~40 minutes\n",
    "\n",
    "# The function returns a tuple with three elements:\n",
    "# 1) A pandas DataFrame with sample ID as index, and three columns aa_haplotype, nucleotide_haplotype and ns_changes\n",
    "# 2) The (3D7) reference amino acid haplotype\n",
    "# 3) The (3D7) reference nucleotide haplotype\n",
    "\n",
    "# Note that the reference haplotypes are probably not required. They were largely intended for debugging/sanity checking purposes\n",
    "# As such the key output is the first element (the DataFrame)\n",
    "\n",
    "####################################\n",
    "# Explanation of column aa_haplotype\n",
    "####################################\n",
    "# If the nucleotide haplotype is homozygous (i.e. there are no heterozygous variants), there will be a single amino acid haplotype\n",
    "# If the nucleotide haplotype is heterozygous (i.e. there is at least one heterozygous variants), there will be two amino acid haplotypes, separated by a comma\n",
    "# Note that in some cases, there could be a heterozygous nucleotide haplotype, both alleles of which give the same amino acid haplotype.\n",
    "# In cases such as this, the amino acid haplotype will be repeated, separated by a comma \n",
    "# If there are multiple mutations that are heterozygous, and cannot be phased, an asterisk (*) will be appended to the aa_haplotype call.\n",
    "# If any variant is missing, aa_haplotype will be '-'.\n",
    "# If the resulting nucleotide haplotype has a length that is not a multiple of 3 (due to indels) the aa_haplotype will be \"!\" to represent a frameshift\n",
    "# In my experience to date, such frameshifts have only be seen as heterozygote calls, i.e. \"<haplotype>,!\" or \"!,<haplotype>\", though would\n",
    "# probably be much more common if setting `include_indels=False`\n",
    "\n",
    "####################################\n",
    "# Explanation of column nucleotide_haplotype\n",
    "####################################\n",
    "# If the nucleotide haplotype is homozygous (i.e. there are no heterozygous variants), there will be a single haplotype\n",
    "# If the nucleotide haplotype is heterozygous (i.e. there is at least one heterozygous variants), there will be two haplotypes, separated by a comma\n",
    "# Nucleotides which match the reference sequence are shown in lower case, whereas nucleotides that differ (SNPs) are shown in upper case\n",
    "# If any variant is missing, nucleotide_haplotype will be '-'.\n",
    "\n",
    "####################################\n",
    "# Explanation of column ns_changes\n",
    "####################################\n",
    "# All non-synonymous mutations will be shown in the form \"<REF AA><AA number><ALT AA>\", e.g. \"K76T\" or \"C580Y\"\n",
    "# Homozygous mutations are shown in upper case and heterozygous in lower case.\n",
    "# If the code finds two distinct amino acid haplotypes, these will be separated by a comma.\n",
    "# If multiple NS mutations are found within an amino acid haplotype, these will be separated by a forward slash (/).\n",
    "# If there are multiple mutations that are heterozygous, and cannot be phased, an asterisk (*) will be appended to the ns_changes call.\n",
    "# Note that this still holds\n",
    "# If there are no NS mutations, ns_changes will be NaN, with the exception of 2 or more unphased synonymous variants, in which case ns_changes will be '*'\n",
    "# This use of '*' to represent multiple unphased synonymous mutations is perhaps undesirable and we might want to review this in the future\n",
    "# Missing variants will be ignored, but any NS mutation elsewhere will still be included. Note the consequence of this is that it is\n",
    "# possible that aa_haplotype will be '-' but ns_changes will have one or more NS mutations reported.\n",
    "# If the resulting nucleotide haplotype has a length that is not a multiple of 3 (due to indels) ns_changes will be \"!\" to represent a frameshift\n",
    "# If any mutations result in a premature stop codon, ns_changes will also be \"!\"\n",
    "\n",
    "####################################\n",
    "# Outline of the code\n",
    "####################################\n",
    "# This function is complex and difficult to follow (I had trouble coming back to it a few months after initially\n",
    "# writing it). In essence this determines the reference nucleotide sequence of the region (ref_sequence), and then\n",
    "# for each sample applies the mutations for the first and second aa_haplotype calls (remembering there are two calls as\n",
    "# we are calling as diploid with GATK). This takes account of complications such as phased het calls and indels changing\n",
    "# the length of the sequence (though indels are not included by default). The reference and two alternative nucleotide\n",
    "# sequences are then converted to amino acid sequences, and mutations detected.\n",
    "\n",
    "####################################\n",
    "# Known issues\n",
    "####################################\n",
    "# Note that if using the default option of `include_indels=False` we shouldn't really see any frameshifts (represented by \"!\" in aa_haplotype),\n",
    "# however, these are occasionally seen in the following two circumstances:\n",
    "# 1) By default we allow the 1bp insertion and deletion at Pf3D7_07_v3:403618 and Pf3D7_07_v3:403622\n",
    "#   (due to default option `positions_to_allow=[403618, 403622]`). These are usually seen together, so result in in-frame changes\n",
    "#   However, in a few cases only one of the indels is called resulting in a frameshift\n",
    "# 2) In some cases a SNP is represented by GATK as REF and ALT alleles longer than 1 but only differing in the first position.\n",
    "#    An example of this in Pf7 is `Pf3D7_08_v3\t549683\t.\tTGG\tCGG,*`. There is a bug in the code that I haven't had chance to look into\n",
    "#    which results in samples with this SNP being represented as a frameshift\n",
    "#\n",
    "# It is possible (and in fact quite common) to have a missing aa_haplotype (represented by \"-\"), but to have one or more mutations in ns_changes.\n",
    "# This was by design and intended specifically for the gene Kelch13 where if we see any non-synonymous mutation between AA 349 and 726\n",
    "# we can assign a resistant phenotype, even if there is missingness in other positions. In a sense aa_haplotype is more conservative than\n",
    "# ns_changes. We might want to review this decision in future\n",
    "#\n",
    "# We have used '*' to represent multiple unphased synonymous mutations in ns_changes.\n",
    "# This is perhaps undesirable and we might want to review this in the future.\n",
    "#\n",
    "# The current function requires a VCF file. This would probably be much quicker to run if using a zarr representation of the VCF\n",
    "\n",
    "def call_haplotype(\n",
    "#     target_name='crt',\n",
    "    chrom='Pf3D7_07_v3',\n",
    "    starts=[403222, 403490, 403938, 404283, 404569, 404764, 404936, 405146, 405334, 405539, 405825, 406017, 406241], # Start positions of each exon\n",
    "    ends=  [403312, 403758, 404110, 404415, 404640, 404839, 405018, 405196, 405390, 405631, 405869, 406071, 406317], # End positions of each exon\n",
    "    strand='+', # Either '+' or '-'\n",
    "    vcf_file_format=vcf_file_format, # Location of VCFs, allowing chrom to vary\n",
    "    samples=None, # The default (None) will call haplotypes for all samples within the VCF\n",
    "    include_indels=False, # If set to True, this will include indels, and hence potentially output sequences of different lengths\n",
    "    positions_with_missingness=[403620, 403621], # If there are missing genotypes at these positions, these will not be treated as missing. This is important in CRT as many samples have missing genotypes at Pf3D7_07_v3:403620-403621 due to way haplotypes are treated by HaplotypeCaller\n",
    "    positions_to_ignore=[], # These positions will be skipped\n",
    "    positions_to_allow=[403618, 403622], # These allows certain indels to be included even if include_indels is False. Important as common CVIET haplotype in CRT 72-76 is treated by HaplotypeCaller as two 1bp indels, rather than 3 SNPs\n",
    "    ref_fasta_fn=ref_genome_fn,\n",
    "    vcf_fn=None, # If supplied, this will overrule what is provided in vcf_file_format\n",
    "    verbose=False,\n",
    "    show_genotypes=False, # If true, these will give very verbose output\n",
    "    use_filter=True, # If true, will only use PASS variants\n",
    "    first_aa=1, # Important to use if, for example, only creating haplotypes for one exon\n",
    "    to_stop=True # This argument is used by Bio translate function. If True, sequences will be truncated at premature stop codons, resulting in shorter AA sequences\n",
    "):\n",
    "    \n",
    "    from pyfasta import Fasta\n",
    "    from Bio.Seq import Seq\n",
    "\n",
    "    \"\"\" This section checks the data types of 'starts' and 'ends' \n",
    "   - Is 'starts' neither a list nor a numpy array? If not, check if its an integer. If it is an integer, convert to list. If not, exit\n",
    "   - Repeat for 'ends' \n",
    "   - Are the converted lists of the same length? Exit if not\n",
    "   - Finally, create a list of exon start positions, relative to the preceeding exon. This is done by calculating diffs between start and ends, \n",
    "   getting the cumsum (minus the last entry) and prefixing with 0. \"\"\"\n",
    "\n",
    "    if not isinstance(starts, list) and not(isinstance(starts, np.ndarray)):\n",
    "        if isinstance(starts, int):\n",
    "            starts = [starts]\n",
    "        else:\n",
    "            sys.exit(\"starts must be a list of integers (or a single integer)\")\n",
    "    if not isinstance(ends, list) and not(isinstance(ends, np.ndarray)):\n",
    "        if isinstance(ends, int):\n",
    "            ends = [ends]\n",
    "        else:\n",
    "            sys.exit(\"ends must be a list of integers (or a single integer)\")\n",
    "    \n",
    "    if len(starts) != len(ends):\n",
    "        sys.exit(\"starts and ends must be the same length\")\n",
    "        \n",
    "    exon_offsets = [0] + list(np.cumsum(np.array(ends)-np.array(starts)+1))[:-1]\n",
    "    \n",
    "    \"\"\" If vcf_fn is not provided, generate a vcf filename based on the value of the chrom variable using the vcf_file_format template.  \"\"\"\n",
    "    if vcf_fn is None:\n",
    "        vcf_fn = vcf_file_format % chrom\n",
    "        \n",
    "    \"\"\"Generate refs: processing a series of exons specified by starts and ends, \n",
    " combining them into a single sequence, and translating that sequence into an amino acid sequence, with proper handling for the strand direction. \"\"\"\n",
    "    ref_sequences = []\n",
    "    for i in np.arange(len(starts)):\n",
    "        start = starts[i]\n",
    "        end = ends[i]\n",
    "        ref_sequences.append(Seq(Fasta(ref_fasta_fn)[chrom][starts[i]-1:ends[i]]))\n",
    "    ref_sequence = Seq(\"\")\n",
    "    for s in ref_sequences:\n",
    "        ref_sequence += s\n",
    "    ref_nucleotide_haplotype = str(ref_sequence)\n",
    "    if strand == '+':\n",
    "        ref_aa_haplotype = str(ref_sequence.translate(to_stop=to_stop))\n",
    "    else:\n",
    "        ref_aa_haplotype = str(ref_sequence.reverse_complement().translate(to_stop=to_stop))\n",
    "\n",
    "    vcf_reader = vcf.Reader(filename=vcf_fn) # Read the vcf file\n",
    "\n",
    "#\"\"\" Set up default samples = None to read in all samples from vcf. Initialise a set of dictionaries\"\"\"\n",
    "    if samples is None:\n",
    "        samples = vcf_reader.samples\n",
    "    sample_sequences = collections.OrderedDict() # will be a set of two sequences to represent the two alt (?) calls for a sample\n",
    "    sample_offsets = collections.OrderedDict()\n",
    "    PID = collections.OrderedDict() # phased IDs\n",
    "    non_phased_het_seen = collections.OrderedDict()\n",
    "    sample_unphaseable = collections.OrderedDict()\n",
    "    aa_haplotype = collections.OrderedDict()\n",
    "    nucleotide_haplotype = collections.OrderedDict()\n",
    "    ns_changes = collections.OrderedDict()\n",
    "    num_missing = collections.OrderedDict()\n",
    "    num_called = collections.OrderedDict()\n",
    "    swap_phasing = collections.OrderedDict()\n",
    "    first_PGT = collections.OrderedDict()\n",
    "    \n",
    "\n",
    "    for sample in samples:\n",
    "        #sample_sequences[sample] = [MutableSeq(ref_sequence), MutableSeq(ref_sequence)] # For using Biopython 1.81\n",
    "        sample_sequences[sample] = [ref_sequence.tomutable(), ref_sequence.tomutable()] # a list with two identical entries.\n",
    "        sample_offsets[sample] = [0, 0]\n",
    "        PID[sample] = ''\n",
    "        non_phased_het_seen[sample] = False\n",
    "        sample_unphaseable[sample] = ''\n",
    "        aa_haplotype[sample] = ''\n",
    "        nucleotide_haplotype[sample] = ''\n",
    "        num_missing[sample] = 0\n",
    "        num_called[sample] = 0\n",
    "        swap_phasing[sample] = False\n",
    "        first_PGT[sample] = ''\n",
    "\n",
    "    for i in np.arange(len(starts)):\n",
    "        start = starts[i]\n",
    "        end = ends[i]\n",
    "        exon_offset = exon_offsets[i]\n",
    "        if verbose:\n",
    "            print(f\"{chrom}:{starts[i]}-{ends[i]}\")\n",
    "        for record in vcf_reader.fetch(chrom, start-1, end): # Find variants in relevant portion of gene\n",
    "            if ( record.POS >= start) and not record.POS in positions_to_ignore: # Have to restrict to position that are at or after start, to ensure indels in previous positions don't mess things up. Note this is a change from previous behaviour\n",
    "                if not use_filter or record.FILTER==[]: # Only use PASS variants, unless use_filter is False\n",
    "                    if verbose:\n",
    "                        print(\"\\n\", record, record.FILTER)\n",
    "                    for sample in samples:\n",
    "                        if show_genotypes:\n",
    "                            print(record.genotype(sample))\n",
    "                        GT = record.genotype(sample)['GT']\n",
    "                        AD = record.genotype(sample)['AD']\n",
    "                        POS = record.POS - start + exon_offset\n",
    "                        if GT == './.' or GT == '.|.': # Note phased empty genotypes (.|.) have appeared in Pf7. These weren't in Pf6!\n",
    "                            if not record.POS in positions_with_missingness:\n",
    "                                num_missing[sample] += 1\n",
    "                        else:\n",
    "                            num_called[sample] += 1\n",
    "                            if GT != '0/0' and GT != '0|0': # Only interested in non-ref genotypes. Hom ref genotypes don't change sequence so can be ignored\n",
    "                                alleles = [record.REF] + record.ALT\n",
    "#                                 first_allele_int = int(GT[0])\n",
    "#                                 second_allele_int = int(GT[2])\n",
    "                                # Find lengths of alleles (to determine which are indels)\n",
    "                                REF_len = len(record.REF)\n",
    "                                first_allele_ALT_len = len(alleles[int(GT[0])]) # not necessarily alts\n",
    "                                second_allele_ALT_len = len(alleles[int(GT[2])])\n",
    "                                if (REF_len==first_allele_ALT_len and REF_len==second_allele_ALT_len) or include_indels or record.POS in positions_to_allow:\n",
    "                                    if (GT[0] != GT[2]) and (alleles[int(GT[0])] != '*') and (alleles[int(GT[2])] != '*'): # heterozygous call at non-spanning deletion\n",
    "                                        if 'PGT' in record.genotype(sample).data._fields:\n",
    "                                            is_unphased = (record.genotype(sample)['PGT'] is None)\n",
    "                                        else:\n",
    "                                            is_unphased = True\n",
    "                                        if non_phased_het_seen[sample]:\n",
    "                                            if verbose:\n",
    "                                                print(\"Sample %s unphased het followed by het\" % sample)\n",
    "                                            sample_unphaseable[sample] = '*'\n",
    "            #                                 genotype[sample] = 'X'\n",
    "                                        if is_unphased:\n",
    "                                            non_phased_het_seen[sample] = True\n",
    "                                        else:\n",
    "#                                             if 'PID' in record.genotype(sample).data._fields: # Note this won't be the case in het spanning deletions\n",
    "                                            if PID[sample] == '':\n",
    "                                                PID[sample] = record.genotype(sample)['PID']\n",
    "                                                first_PGT[sample] = record.genotype(sample)['PGT']\n",
    "                                                if AD[int(GT[2])] > AD[int(GT[0])]:\n",
    "                                                    swap_phasing[sample] = True\n",
    "                                            else:\n",
    "                                                if record.genotype(sample)['PID'] != PID[sample]:\n",
    "                                                    if verbose:\n",
    "                                                        print(\"Sample %s two PIDs\" % sample)\n",
    "                                                    sample_unphaseable[sample] = '*'\n",
    "            #                                         genotype[sample] = '?'\n",
    "#                                             if record.genotype(sample)['PGT'] == '1|0':\n",
    "#                                                 GT = GT[2] + '/' + GT[0]\n",
    "                                        if verbose:\n",
    "                                            print(f\"AD={AD}, GT={GT}\")\n",
    "                                        # Here we use phased genotype (PGT) if in same phasing group as first het\n",
    "                                        if (\n",
    "                                            'PID' in record.genotype(sample).data._fields\n",
    "                                            and ( PID[sample] == record.genotype(sample)['PID'] )\n",
    "                                        ): # In phasing group with first het call\n",
    "                                            if verbose:\n",
    "                                                print(f\"PID: {PID[sample]}, GT: {GT}, PGT: {record.genotype(sample)['PGT']}\")\n",
    "                                            if first_PGT[sample] == record.genotype(sample)['PGT']:\n",
    "                                                if swap_phasing[sample]:\n",
    "                                                    GT = GT[2] + '/' + GT[0]\n",
    "                                            else:\n",
    "                                                if not swap_phasing[sample]:\n",
    "                                                    GT = GT[2] + '/' + GT[0]\n",
    "#                                             GT = record.genotype(sample)['PGT']\n",
    "                                        # Else we ensure we are taking the majority haplotype as the first allele\n",
    "                                        elif AD[int(GT[2])] > AD[int(GT[0])]:\n",
    "                                            GT = GT[2] + '/' + GT[0]\n",
    "                                            if verbose:\n",
    "                                                print(f\"AD={AD}, GT={GT}\")\n",
    "\n",
    "                                for i, sample_offset in enumerate(sample_offsets[sample]):\n",
    "                                    alleles = [record.REF] + record.ALT\n",
    "                                    GTint = int(GT[i*2])\n",
    "                                    REF_len = len(record.REF)\n",
    "                                    ALT_len = len(alleles[GTint])\n",
    "                                    if(verbose):\n",
    "                                        print(f\"before: i={i}, offset={sample_offset}, alleles, GT={GTint}, REF={REF_len}, ALT={ALT_len}, {sample_sequences[sample][i]}\")\n",
    "                                    if GTint != 0 and alleles[GTint] != '*' and (REF_len==ALT_len or include_indels or record.POS in positions_to_allow):\n",
    "                                        sample_sequences[sample][i][POS+sample_offsets[sample][i]:(POS+sample_offsets[sample][i]+REF_len)] = alleles[GTint]\n",
    "                                        sample_offsets[sample][i] = sample_offset + (len(alleles[GTint]) - len(record.REF))\n",
    "                                        if(verbose):\n",
    "                                            print(f\"after : i={i}, offset={sample_offset}, alleles, GT={GTint}, REF={REF_len}, ALT={ALT_len}, {sample_sequences[sample][i]}\")\n",
    "\n",
    "    for sample in samples:\n",
    "        if(show_genotypes):\n",
    "            print(sample_sequences[sample])\n",
    "        ###########################\n",
    "        # populate ns_changes\n",
    "        ###########################\n",
    "        if strand == '+':\n",
    "            aa_0 = (\n",
    "                str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "            aa_1 = (\n",
    "                str(sample_sequences[sample][1].toseq().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "        else:\n",
    "            aa_0 = (\n",
    "                str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "            aa_1 = (\n",
    "                str(sample_sequences[sample][1].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                else '!'\n",
    "            )\n",
    "#         if aa_0 != '!' and aa_1 != '!' # This was what was used in Pf6\n",
    "        if aa_0 != '!' and aa_1 != '!' and (len(aa_0) == len(ref_aa_haplotype)) and (len(aa_1) == len(ref_aa_haplotype)): # In Pf7, we have some in-frame indels, hence we also need to check for these, and give \"!\" if indels exist\n",
    "            if aa_0 == aa_1:\n",
    "                ns_changes_list = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "                    if aa_0[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_0[i]))\n",
    "                ns_changes[sample] = \"/\".join(ns_changes_list)\n",
    "            else:\n",
    "                ns_changes_list_0 = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "                    if aa_0[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list_0.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_0[i]))\n",
    "                ns_changes_0 = \"/\".join(ns_changes_list_0)\n",
    "                ns_changes_list_1 = []\n",
    "                for i in range(len(ref_aa_haplotype)):\n",
    "#                     print(len(ref_aa_haplotype), len(aa_1))\n",
    "                    if aa_1[i] != ref_aa_haplotype[i]:\n",
    "                        ns_changes_list_1.append(\"%s%d%s\" % (ref_aa_haplotype[i], i+first_aa, aa_1[i]))\n",
    "                ns_changes_1 = \"/\".join(ns_changes_list_1)\n",
    "\n",
    "                if ns_changes_0 == '':\n",
    "                    ns_changes[sample] = ns_changes_1.lower()\n",
    "                elif ns_changes_1 == '':\n",
    "                    ns_changes[sample] = ns_changes_0.lower()\n",
    "                else:\n",
    "                    ns_changes[sample] = (\",\".join([ns_changes_0, ns_changes_1])).lower()\n",
    "        else:\n",
    "            if verbose:\n",
    "                print(aa_0, aa_1, ref_aa_haplotype)\n",
    "            ns_changes[sample] = '!'\n",
    "\n",
    "        ################################################\n",
    "        # populate nucleotide_haplotype and aa_haplotype\n",
    "        ################################################\n",
    "        if str(sample_sequences[sample][0]).upper() == str(sample_sequences[sample][1]).upper():\n",
    "            nucleotide_haplotype[sample] = str(sample_sequences[sample][0].toseq())\n",
    "            if strand == '+':\n",
    "                aa_haplotype[sample] = ( str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                                    if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                                    else '!' )\n",
    "            else:\n",
    "                aa_haplotype[sample] = ( str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                                    if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                                    else '!' )\n",
    "        else:\n",
    "            nucleotide_haplotype[sample] = \"%s,%s\" % ( str(sample_sequences[sample][0].toseq()), str(sample_sequences[sample][1].toseq()) )\n",
    "            if strand == '+':\n",
    "                aa_haplotype[sample] = \"%s,%s\" % (\n",
    "                    ( str(sample_sequences[sample][0].toseq().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                     else '!' ),\n",
    "                    ( str(sample_sequences[sample][1].toseq().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                     else '!' )\n",
    "                )\n",
    "            else:\n",
    "                aa_haplotype[sample] = \"%s,%s\" % (\n",
    "                    ( str(sample_sequences[sample][0].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][0]) % 3 == 0\n",
    "                     else '!' ),\n",
    "                    ( str(sample_sequences[sample][1].toseq().reverse_complement().translate(to_stop=to_stop))\n",
    "                     if len(sample_sequences[sample][1]) % 3 == 0\n",
    "                     else '!' )\n",
    "                )\n",
    "    \n",
    "        # Add asterisk to end if sample is unphaseable\n",
    "        ns_changes[sample] += sample_unphaseable[sample]\n",
    "        aa_haplotype[sample] += sample_unphaseable[sample]\n",
    "        nucleotide_haplotype[sample] += sample_unphaseable[sample]\n",
    "\n",
    "        # Set ns_changes output to missing ('-') if no NS changes and at least one missing genotype call in the region\n",
    "        if (ns_changes[sample] == '' or ns_changes[sample] == '*') and num_missing[sample] > 0:\n",
    "            ns_changes[sample] = '-'\n",
    "                \n",
    "        # Set sequence output to missing ('-') if any missing genotype call in the gene\n",
    "        if num_missing[sample] > 0:\n",
    "            aa_haplotype[sample] = '-'\n",
    "            nucleotide_haplotype[sample] = '-'\n",
    "    \n",
    "    df_all_haplotypes = pd.DataFrame(\n",
    "        {\n",
    "            'aa_haplotype': pd.Series(aa_haplotype),\n",
    "            'nucleotide_haplotype': pd.Series(nucleotide_haplotype),\n",
    "            'ns_changes': pd.Series(ns_changes),\n",
    "        }\n",
    "    )\n",
    "    call_haps_pickle_fn = f'call_haplotypes_data_{gene_id}.pkl'\n",
    "    with open(call_haps_pickle_fn, 'wb') as file:\n",
    "        call_haplotypes_data = (df_all_haplotypes, ref_aa_haplotype, ref_nucleotide_haplotype)\n",
    "        pickle.dump(call_haplotypes_data, file)\n",
    "        \n",
    "#     return(df_haplotypes, ref_aa_haplotype, ref_nucleotide_haplotype)\n",
    "    return(\n",
    "        {\n",
    "            'df_all_haplotypes':df_all_haplotypes,\n",
    "            'ref_aa_haplotype':ref_aa_haplotype,\n",
    "            'ref_nucleotide_haplotype':ref_nucleotide_haplotype,\n",
    "        }\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_plot_data(\n",
    "    gene_id='PF3D7_0709000',\n",
    "    haplotype_calls=None,\n",
    "    samples_fn=samples_fn,\n",
    "    expected_aa_len=None,\n",
    "#     populations = ['SA', 'AF-W', 'AF-C', 'AF-NE', 'AF-E', 'AS-S-E', 'AS-S-FE', 'AS-SE-W', 'AS-SE-E', 'OC-NG', ],\n",
    "    populations = ['AF-W', 'AF-C', 'AF-NE', 'AF-E', ],\n",
    "    samples_to_show = {\n",
    "        'PG0563-C': '7G8',\n",
    "        'PG0564-C': 'GB4',\n",
    "        'PG0565-C': 'HB3',\n",
    "        'PG0566-C': 'Dd2',\n",
    "        'PG0567-C': '3D7',\n",
    "        'PG0568-C': 'IT',\n",
    "    },\n",
    "    sample_to_use_as_background = 'PG0566-C', # Dd2\n",
    "    verbose=True,\n",
    "    **kwargs\n",
    "):\n",
    "#     print(gene_id)\n",
    "    if gene_id is not None:\n",
    "        cds_results = determine_cds(gene_id)\n",
    "#     print(cds_results['gene_name'])\n",
    "    \n",
    "        if haplotype_calls is None:\n",
    "            haplotype_calls = call_haplotype(\n",
    "                chrom=cds_results['chrom'],\n",
    "                starts=cds_results['starts'],\n",
    "                ends=cds_results['ends'],\n",
    "                strand=cds_results['strand'],\n",
    "                **kwargs\n",
    "            )\n",
    "    else:\n",
    "        haplotype_calls = call_haplotype(\n",
    "            **kwargs\n",
    "        )\n",
    "        \n",
    "    \n",
    "    df_samples = pd.read_csv(samples_fn, sep='\\t', low_memory=False, index_col=0)\n",
    "    if verbose:\n",
    "        print(f\"Read in {df_samples.shape[0]:,} samples, of which {np.count_nonzero(df_samples['QC pass']):,} passed QC\")\n",
    "\n",
    "    df_join = df_samples.join(haplotype_calls['df_all_haplotypes'])\n",
    "    \n",
    "    if expected_aa_len is None:\n",
    "        expected_aa_len = len(haplotype_calls['ref_aa_haplotype'])\n",
    "    \n",
    "    # check you dont have two haplotypes. Clonal sample aa haplotype will == reference. Mixed will have two+ strings separated by comma.\n",
    "    df_clonal = df_join.loc[\n",
    "        (\n",
    "            ( df_join['QC pass'] == True )\n",
    "            & ( df_join['aa_haplotype'].astype(str).apply(len) == expected_aa_len )\n",
    "        ) | ( df_join['Study'] == '1153-PF-Pf3KLAB-KWIATKOWSKI' ),\n",
    "        [\n",
    "            'aa_haplotype',\n",
    "            'nucleotide_haplotype',\n",
    "            'ns_changes',\n",
    "            'Population',\n",
    "        ]\n",
    "    ]\n",
    "    \n",
    "    if df_clonal.shape[0] == 0:\n",
    "        most_common_length = df_join['aa_haplotype'].astype(str).apply(len).value_counts().index[0]\n",
    "        raise Exception(f\"There were no samples with AA haplotype length = {expected_aa_len}. Most common length was {most_common_length}\")\n",
    "    \n",
    "    df_clonal['ns_changes'].fillna('', inplace=True)\n",
    "    df_clonal['number_of_mutations'] = df_clonal['ns_changes'].apply(lambda x: len(x.split('/')))\n",
    "    df_clonal.loc[df_clonal['ns_changes'] == '', 'number_of_mutations'] = 0\n",
    "    if verbose:\n",
    "        print(f\"Found {df_clonal.shape[0]:,} clonal haplotypes in QC pass samples\")\n",
    "\n",
    "    grouping_columns = [\n",
    "        'number_of_mutations', 'ns_changes',\n",
    "    #     'haplotype_name',\n",
    "    ]\n",
    "    ascending=[True]*len(grouping_columns)\n",
    "    ascending.append(False)\n",
    "\n",
    "    def n_agg(x):\n",
    "        names = collections.OrderedDict()\n",
    "        for population in populations:\n",
    "            names[population] = np.count_nonzero( (x['Population'] == population) )\n",
    "        return pd.Series(names)\n",
    "\n",
    "    df_haplotypes = df_clonal.groupby(grouping_columns).apply(n_agg)\n",
    "    df_haplotypes['Total'] = df_haplotypes.sum(axis=1)\n",
    "    df_haplotypes.reset_index(inplace=True)\n",
    "    df_haplotypes.sort_values(grouping_columns + ['Total'], ascending=ascending, inplace=True)\n",
    "    df_haplotypes['ns_changes_list'] = df_haplotypes['ns_changes'].apply(lambda x: x.split('/'))\n",
    "    df_haplotypes.sort_values('Total', ascending=False, inplace=True)\n",
    "    # Determine which samples to show have which haplotypes\n",
    "    df_haplotypes['sample_names'] = ''\n",
    "    for sample in samples_to_show.keys():\n",
    "        ns_changes = df_clonal.loc[sample, 'ns_changes']\n",
    "        if df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'].values == '':\n",
    "            df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] = samples_to_show[sample]\n",
    "        else:\n",
    "            df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] = (\n",
    "                df_haplotypes.loc[df_haplotypes['ns_changes']==ns_changes, 'sample_names'] + '\\n' + samples_to_show[sample]\n",
    "            )\n",
    "    if verbose:\n",
    "        print(f\"Found {df_haplotypes.shape[0]:,} unique haplotypes in QC pass samples\")\n",
    "        \n",
    "    # Determine background mutations\n",
    "    if sample_to_use_as_background in df_clonal.index:\n",
    "        background_ns_changes = df_clonal.loc[sample_to_use_as_background, 'ns_changes']\n",
    "        \n",
    "    if gene_id is not None:\n",
    "        gene_name = cds_results['gene_name']\n",
    "    else:\n",
    "        gene_name = None\n",
    "    \n",
    "    prep_plot_pickle_fn = f'prep_plot_data_{gene_id}.pkl'\n",
    "    with open(prep_plot_pickle_fn, 'wb') as file:\n",
    "        plot_data = (df_haplotypes, df_clonal, df_join, background_ns_changes, haplotype_calls, gene_name, df_samples)\n",
    "        pickle.dump(plot_data, file)\n",
    "    \n",
    "\n",
    "    return(\n",
    "        {\n",
    "            'df_haplotypes':df_haplotypes,\n",
    "            'df_clonal':df_clonal,\n",
    "            'df_join':df_join,\n",
    "            'background_ns_changes':background_ns_changes, \n",
    "            'haplotype_calls':haplotype_calls, \n",
    "            'gene_name':gene_name, \n",
    "            'df_samples':df_samples,\n",
    "        }\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_of_interest = collections.OrderedDict()\n",
    "\n",
    "genes_of_interest['PF3D7_0417200'] = 'DHFR'\n",
    "genes_of_interest['PF3D7_0523000'] = 'MDR1'\n",
    "genes_of_interest['PF3D7_0709000'] = 'CRT'\n",
    "genes_of_interest['PF3D7_0810800'] = 'DHPS'\n",
    "genes_of_interest['PF3D7_1224000'] = 'GCH1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PF3D7_0417200 DHFR\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 13,781 clonal haplotypes in QC pass samples\n",
      "Found 30 unique haplotypes in QC pass samples\n",
      "\n",
      "PF3D7_0523000 MDR1\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 10,239 clonal haplotypes in QC pass samples\n",
      "Found 323 unique haplotypes in QC pass samples\n",
      "\n",
      "PF3D7_0709000 CRT\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 12,405 clonal haplotypes in QC pass samples\n",
      "Found 101 unique haplotypes in QC pass samples\n",
      "\n",
      "PF3D7_0810800 DHPS\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 12,533 clonal haplotypes in QC pass samples\n",
      "Found 78 unique haplotypes in QC pass samples\n",
      "\n",
      "PF3D7_1224000 GCH1\n",
      "Read in 20,864 samples, of which 16,203 passed QC\n",
      "Found 14,886 clonal haplotypes in QC pass samples\n",
      "Found 97 unique haplotypes in QC pass samples\n",
      "\n",
      "CPU times: user 27min 15s, sys: 5.41 s, total: 27min 21s\n",
      "Wall time: 27min 29s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "results=collections.OrderedDict()\n",
    "for gene_id in genes_of_interest:\n",
    "    print(gene_id, genes_of_interest[gene_id])\n",
    "    prepare_plot_data(gene_id=gene_id)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the call haplotypes pickle file for reading \n",
    "gene_id = 'PF3D7_0417200'\n",
    "#gene_id = 'PF3D7_0523000'\n",
    "#gene_id = 'PF3D7_0709000' \n",
    "#gene_id = 'PF3D7_0810800'\n",
    "#gene_id = 'PF3D7_1224000' \n",
    "\n",
    "with open(f'call_haplotypes_data_{gene_id}.pkl', 'rb') as file:\n",
    "    loaded_haps_data = pickle.load(file)\n",
    "\n",
    "# Access the data using indexes of the loaded data tuple\n",
    "df_all_haplotypes = loaded_haps_data[0]\n",
    "ref_aa_haplotype = loaded_haps_data[1]\n",
    "ref_nucleotide_haplotype = loaded_haps_data[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_haplotype</th>\n",
       "      <th>nucleotide_haplotype</th>\n",
       "      <th>ns_changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>s108n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               aa_haplotype  \\\n",
       "FP0008-C  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0009-C  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "\n",
       "                                       nucleotide_haplotype       ns_changes  \n",
       "FP0008-C  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...            s108n  \n",
       "FP0009-C  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_haplotypes.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVLPWKCNSLDMKYFCAVTTYVNESKYEKLKYKRCKYLNKETVDNVNDMPNSKKLQNVVVMGRTSWESIPKKFKPLSNRINVILSRTLKKEDFDEDVYIINKVEDLIVLLGKLNYYKCFIIGGSVVYQEFLEKKLIKKIYFTRINSTYECDVFFPEINENEYQIISVSDVYTSNNTTLDFIIYKKTNNKMLNEQNCIKGEEKNNDMPLKNDDKDTCHMKKLTEFYKNVDKYKINYENDDDDEEEDDFVYFNFNKEKEEKNKNSIHPNDFQIYNSLKYKYHPEYQYLNIIYDIMMNGNKQSDRTGVGVLSKFGYIMKFDLSQYFPLLTTKKLFLRGIIEELLWFIRGETNGNTLLNKNVRIWEANGTREFLDNRKLFHREVNDLGPIYGFQWRHFGAEYTNMYDNYENKGVDQLKNIINLIKNDPTSRRILLCAWNVKDLDQMALPPCHILCQFYVFDGKLSCIMYQRSCDLGLGVPFNIASYSIFTHMIAQVCNLQPAQFIHVLGNAHVYNNHIDSLKIQLNRIPYPFPTLKLNPDIKNIEDFTISDFTIQNYVHHEKISMDMAA\n",
      "atgatggaacaagtctgcgacgttttcgatatttatgccatatgtgcatgttgtaaggttgaaagcaaaaatgaggggaaaaaaaatgaggtttttaataactacacatttagaggtctaggaaataaaggagtattaccatggaaatgtaattccctagatatgaaatatttttgtgcagttacaacatatgtgaatgaatcaaaatatgaaaaattgaaatataagagatgtaaatatttaaacaaagaaactgtggataatgtaaatgatatgcctaattctaaaaaattacaaaatgttgtagttatgggaagaacaagctgggaaagcattccaaaaaaatttaaacctttaagcaataggataaatgttatattgtctagaaccttaaaaaaagaagattttgatgaagatgtttatatcattaacaaagttgaagatctaatagttttacttgggaaattaaattactataaatgttttattataggaggttccgttgtttatcaagaatttttagaaaagaaattaataaaaaaaatatattttactagaataaatagtacatatgaatgtgatgtattttttccagaaataaatgaaaatgagtatcaaattatttctgttagcgatgtatatactagtaacaatacaacattggattttatcatttataagaaaacgaataataaaatgttaaatgaacaaaattgtataaaaggagaagaaaaaaataatgatatgcctttaaagaatgatgacaaagatacatgtcatatgaaaaaattaacagaattttacaaaaatgtagacaaatataaaattaattatgaaaatgatgatgatgatgaagaagaagatgattttgtttattttaattttaataaagaaaaagaagagaaaaataaaaattctatacatccaaatgattttcaaatatataatagcttgaaatataaatatcatcctgaataccaatatttaaatattatttatgatattatgatgaatggaaataaacaaagtgatcgaacgggagtaggtgttttaagtaaattcggatatattatgaaatttgatttaagtcaatatttcccattattaactacgaagaaattatttttaagaggaattattgaagaattgctttggtttattagaggagaaacaaatggtaatacgttgttaaataagaatgtaaggatatgggaagctaatggtactagggaatttttagataatagaaaattatttcatagagaagttaacgatttaggacctatttatggttttcaatggagacatttcggtgctgaatatacaaatatgtatgataattatgaaaataaaggagtggatcaattaaaaaatataataaatttaattaaaaatgatcctacaagtagaagaattcttttgtgtgcatggaatgtaaaagatcttgaccaaatggcattacctccttgtcatattttatgtcagttttatgttttcgatgggaaattatcatgtattatgtatcaaagatcatgtgatttagggctaggagtaccttttaatattgcttcttattctatttttactcatatgattgcacaagtctgtaatttgcaacctgcgcagttcatacacgttttaggaaatgcacatgtttataataatcacattgatagtttaaaaattcaacttaacagaataccctatccattcccaacacttaaattaaatccagatattaaaaatattgaagattttacaatttcggattttacaatacaaaattatgttcatcatgaaaaaatttcaatggatatggctgcttaa\n"
     ]
    }
   ],
   "source": [
    "print(ref_aa_haplotype)\n",
    "print(ref_nucleotide_haplotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the plot data pickle file for reading \n",
    "gene_id = 'PF3D7_0417200'\n",
    "#gene_id = 'PF3D7_0523000'\n",
    "#gene_id = 'PF3D7_0709000' \n",
    "#gene_id = 'PF3D7_0810800'\n",
    "#gene_id = 'PF3D7_1224000' \n",
    "\n",
    "with open(f'prep_plot_data_{gene_id}.pkl', 'rb') as file:\n",
    "    loaded_plot_data = pickle.load(file)\n",
    "\n",
    "# Access the data using indexes of the loaded data tuple\n",
    "df_haplotypes = loaded_plot_data[0]\n",
    "df_clonal = loaded_plot_data[1]\n",
    "df_join = loaded_plot_data[2]\n",
    "background_ns_changes = loaded_plot_data[3]\n",
    "haplotype_calls = loaded_plot_data[4]\n",
    "gene_name = loaded_plot_data[5]\n",
    "df_samples = loaded_plot_data[6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_of_mutations</th>\n",
       "      <th>ns_changes</th>\n",
       "      <th>AF-W</th>\n",
       "      <th>AF-C</th>\n",
       "      <th>AF-NE</th>\n",
       "      <th>AF-E</th>\n",
       "      <th>Total</th>\n",
       "      <th>ns_changes_list</th>\n",
       "      <th>sample_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>3905</td>\n",
       "      <td>373</td>\n",
       "      <td>100</td>\n",
       "      <td>1085</td>\n",
       "      <td>5463</td>\n",
       "      <td>[N51I, C59R, S108N]</td>\n",
       "      <td>Dd2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>661</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>47</td>\n",
       "      <td>713</td>\n",
       "      <td>[]</td>\n",
       "      <td>GB4\\n3D7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>C59R/S108N</td>\n",
       "      <td>260</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>75</td>\n",
       "      <td>339</td>\n",
       "      <td>[C59R, S108N]</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>N51I/S108N</td>\n",
       "      <td>52</td>\n",
       "      <td>59</td>\n",
       "      <td>45</td>\n",
       "      <td>99</td>\n",
       "      <td>255</td>\n",
       "      <td>[N51I, S108N]</td>\n",
       "      <td>7G8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>S108N</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>[S108N]</td>\n",
       "      <td>HB3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    number_of_mutations       ns_changes  AF-W  AF-C  AF-NE  AF-E  Total  \\\n",
       "19                    3  N51I/C59R/S108N  3905   373    100  1085   5463   \n",
       "0                     0                    661     2      3    47    713   \n",
       "11                    2       C59R/S108N   260     3      1    75    339   \n",
       "12                    2       N51I/S108N    52    59     45    99    255   \n",
       "7                     1            S108N    17     2      2     4     25   \n",
       "\n",
       "        ns_changes_list sample_names  \n",
       "19  [N51I, C59R, S108N]          Dd2  \n",
       "0                    []     GB4\\n3D7  \n",
       "11        [C59R, S108N]               \n",
       "12        [N51I, S108N]          7G8  \n",
       "7               [S108N]          HB3  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_haplotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_haplotype</th>\n",
       "      <th>nucleotide_haplotype</th>\n",
       "      <th>ns_changes</th>\n",
       "      <th>Population</th>\n",
       "      <th>number_of_mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0011-CW</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0012-CW</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0013-CW</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0015-C</th>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                aa_haplotype  \\\n",
       "Sample                                                         \n",
       "FP0009-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0011-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0012-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0013-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0015-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "\n",
       "                                        nucleotide_haplotype       ns_changes  \\\n",
       "Sample                                                                          \n",
       "FP0009-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N   \n",
       "FP0011-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N   \n",
       "FP0012-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N   \n",
       "FP0013-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N   \n",
       "FP0015-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N   \n",
       "\n",
       "          Population  number_of_mutations  \n",
       "Sample                                     \n",
       "FP0009-C        AF-W                    3  \n",
       "FP0011-CW       AF-W                    3  \n",
       "FP0012-CW       AF-W                    3  \n",
       "FP0013-CW       AF-W                    3  \n",
       "FP0015-C        AF-W                    3  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clonal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>Population</th>\n",
       "      <th>% callable</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf6</th>\n",
       "      <th>aa_haplotype</th>\n",
       "      <th>nucleotide_haplotype</th>\n",
       "      <th>ns_changes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>82.16</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>s108n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>88.85</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>86.46</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0011-CW</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889624</td>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>86.35</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0012-CW</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889627</td>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>89.74</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "      <td>MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...</td>\n",
       "      <td>atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...</td>\n",
       "      <td>N51I/C59R/S108N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Study     Country   Admin level 1  Country latitude  \\\n",
       "Sample                                                                       \n",
       "FP0008-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0009-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0011-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0012-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "\n",
       "           Country longitude  Admin level 1 latitude  Admin level 1 longitude  \\\n",
       "Sample                                                                          \n",
       "FP0008-C          -10.337093               16.565426                -9.832345   \n",
       "FP0009-C          -10.337093               16.565426                -9.832345   \n",
       "FP0010-CW         -10.337093               16.565426                -9.832345   \n",
       "FP0011-CW         -10.337093               16.565426                -9.832345   \n",
       "FP0012-CW         -10.337093               16.565426                -9.832345   \n",
       "\n",
       "             Year         ENA All samples same case Population  % callable  \\\n",
       "Sample                                                                       \n",
       "FP0008-C   2014.0  ERR1081237              FP0008-C       AF-W       82.16   \n",
       "FP0009-C   2014.0  ERR1081238              FP0009-C       AF-W       88.85   \n",
       "FP0010-CW  2014.0  ERR2889621             FP0010-CW       AF-W       86.46   \n",
       "FP0011-CW  2014.0  ERR2889624             FP0011-CW       AF-W       86.35   \n",
       "FP0012-CW  2014.0  ERR2889627             FP0012-CW       AF-W       89.74   \n",
       "\n",
       "           QC pass Exclusion reason Sample type  Sample was in Pf6  \\\n",
       "Sample                                                               \n",
       "FP0008-C      True     Analysis_set        gDNA               True   \n",
       "FP0009-C      True     Analysis_set        gDNA               True   \n",
       "FP0010-CW     True     Analysis_set        sWGA              False   \n",
       "FP0011-CW     True     Analysis_set        sWGA              False   \n",
       "FP0012-CW     True     Analysis_set        sWGA              False   \n",
       "\n",
       "                                                aa_haplotype  \\\n",
       "Sample                                                         \n",
       "FP0008-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0009-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0010-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0011-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0012-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "\n",
       "                                        nucleotide_haplotype       ns_changes  \n",
       "Sample                                                                         \n",
       "FP0008-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...            s108n  \n",
       "FP0009-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0010-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0011-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0012-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_join.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'N51I/C59R/S108N'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "background_ns_changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['df_all_haplotypes', 'ref_aa_haplotype', 'ref_nucleotide_haplotype'])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "haplotype_calls.keys() # This is the output of call_haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values([                                                aa_haplotype  \\\n",
       "FP0008-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0009-C   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0010-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0011-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "FP0012-CW  MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "...                                                      ...   \n",
       "SPT43399   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "SPT43400   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "SPT43401   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "SPT43403                                                   -   \n",
       "SPT43404   MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVL...   \n",
       "\n",
       "                                        nucleotide_haplotype       ns_changes  \n",
       "FP0008-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...            s108n  \n",
       "FP0009-C   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0010-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0011-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "FP0012-CW  atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...  N51I/C59R/S108N  \n",
       "...                                                      ...              ...  \n",
       "SPT43399   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...                   \n",
       "SPT43400   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...                   \n",
       "SPT43401   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...                   \n",
       "SPT43403                                                   -                -  \n",
       "SPT43404   atgatggaacaagtctgcgacgttttcgatatttatgccatatgtg...                   \n",
       "\n",
       "[20864 rows x 3 columns], 'MMEQVCDVFDIYAICACCKVESKNEGKKNEVFNNYTFRGLGNKGVLPWKCNSLDMKYFCAVTTYVNESKYEKLKYKRCKYLNKETVDNVNDMPNSKKLQNVVVMGRTSWESIPKKFKPLSNRINVILSRTLKKEDFDEDVYIINKVEDLIVLLGKLNYYKCFIIGGSVVYQEFLEKKLIKKIYFTRINSTYECDVFFPEINENEYQIISVSDVYTSNNTTLDFIIYKKTNNKMLNEQNCIKGEEKNNDMPLKNDDKDTCHMKKLTEFYKNVDKYKINYENDDDDEEEDDFVYFNFNKEKEEKNKNSIHPNDFQIYNSLKYKYHPEYQYLNIIYDIMMNGNKQSDRTGVGVLSKFGYIMKFDLSQYFPLLTTKKLFLRGIIEELLWFIRGETNGNTLLNKNVRIWEANGTREFLDNRKLFHREVNDLGPIYGFQWRHFGAEYTNMYDNYENKGVDQLKNIINLIKNDPTSRRILLCAWNVKDLDQMALPPCHILCQFYVFDGKLSCIMYQRSCDLGLGVPFNIASYSIFTHMIAQVCNLQPAQFIHVLGNAHVYNNHIDSLKIQLNRIPYPFPTLKLNPDIKNIEDFTISDFTIQNYVHHEKISMDMAA', 'atgatggaacaagtctgcgacgttttcgatatttatgccatatgtgcatgttgtaaggttgaaagcaaaaatgaggggaaaaaaaatgaggtttttaataactacacatttagaggtctaggaaataaaggagtattaccatggaaatgtaattccctagatatgaaatatttttgtgcagttacaacatatgtgaatgaatcaaaatatgaaaaattgaaatataagagatgtaaatatttaaacaaagaaactgtggataatgtaaatgatatgcctaattctaaaaaattacaaaatgttgtagttatgggaagaacaagctgggaaagcattccaaaaaaatttaaacctttaagcaataggataaatgttatattgtctagaaccttaaaaaaagaagattttgatgaagatgtttatatcattaacaaagttgaagatctaatagttttacttgggaaattaaattactataaatgttttattataggaggttccgttgtttatcaagaatttttagaaaagaaattaataaaaaaaatatattttactagaataaatagtacatatgaatgtgatgtattttttccagaaataaatgaaaatgagtatcaaattatttctgttagcgatgtatatactagtaacaatacaacattggattttatcatttataagaaaacgaataataaaatgttaaatgaacaaaattgtataaaaggagaagaaaaaaataatgatatgcctttaaagaatgatgacaaagatacatgtcatatgaaaaaattaacagaattttacaaaaatgtagacaaatataaaattaattatgaaaatgatgatgatgatgaagaagaagatgattttgtttattttaattttaataaagaaaaagaagagaaaaataaaaattctatacatccaaatgattttcaaatatataatagcttgaaatataaatatcatcctgaataccaatatttaaatattatttatgatattatgatgaatggaaataaacaaagtgatcgaacgggagtaggtgttttaagtaaattcggatatattatgaaatttgatttaagtcaatatttcccattattaactacgaagaaattatttttaagaggaattattgaagaattgctttggtttattagaggagaaacaaatggtaatacgttgttaaataagaatgtaaggatatgggaagctaatggtactagggaatttttagataatagaaaattatttcatagagaagttaacgatttaggacctatttatggttttcaatggagacatttcggtgctgaatatacaaatatgtatgataattatgaaaataaaggagtggatcaattaaaaaatataataaatttaattaaaaatgatcctacaagtagaagaattcttttgtgtgcatggaatgtaaaagatcttgaccaaatggcattacctccttgtcatattttatgtcagttttatgttttcgatgggaaattatcatgtattatgtatcaaagatcatgtgatttagggctaggagtaccttttaatattgcttcttattctatttttactcatatgattgcacaagtctgtaatttgcaacctgcgcagttcatacacgttttaggaaatgcacatgtttataataatcacattgatagtttaaaaattcaacttaacagaataccctatccattcccaacacttaaattaaatccagatattaaaaatattgaagattttacaatttcggattttacaatacaaaattatgttcatcatgaaaaaatttcaatggatatggctgcttaa'])"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "haplotype_calls.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DHFR-TS'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Country</th>\n",
       "      <th>Admin level 1</th>\n",
       "      <th>Country latitude</th>\n",
       "      <th>Country longitude</th>\n",
       "      <th>Admin level 1 latitude</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>Population</th>\n",
       "      <th>% callable</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP0008-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>82.16</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0009-C</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>88.85</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP0010-CW</th>\n",
       "      <td>1147-PF-MR-CONWAY</td>\n",
       "      <td>Mauritania</td>\n",
       "      <td>Hodh el Gharbi</td>\n",
       "      <td>20.265149</td>\n",
       "      <td>-10.337093</td>\n",
       "      <td>16.565426</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>86.46</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Study     Country   Admin level 1  Country latitude  \\\n",
       "Sample                                                                       \n",
       "FP0008-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0009-C   1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "FP0010-CW  1147-PF-MR-CONWAY  Mauritania  Hodh el Gharbi         20.265149   \n",
       "\n",
       "           Country longitude  Admin level 1 latitude  Admin level 1 longitude  \\\n",
       "Sample                                                                          \n",
       "FP0008-C          -10.337093               16.565426                -9.832345   \n",
       "FP0009-C          -10.337093               16.565426                -9.832345   \n",
       "FP0010-CW         -10.337093               16.565426                -9.832345   \n",
       "\n",
       "             Year         ENA All samples same case Population  % callable  \\\n",
       "Sample                                                                       \n",
       "FP0008-C   2014.0  ERR1081237              FP0008-C       AF-W       82.16   \n",
       "FP0009-C   2014.0  ERR1081238              FP0009-C       AF-W       88.85   \n",
       "FP0010-CW  2014.0  ERR2889621             FP0010-CW       AF-W       86.46   \n",
       "\n",
       "           QC pass Exclusion reason Sample type  Sample was in Pf6  \n",
       "Sample                                                              \n",
       "FP0008-C      True     Analysis_set        gDNA               True  \n",
       "FP0009-C      True     Analysis_set        gDNA               True  \n",
       "FP0010-CW     True     Analysis_set        sWGA              False  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_samples.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
